# Collaborative Filtering implementation in Keras

To demonstrate the use of modern Machine learning libraries with Automatic Differenciation we re-implement 
the Collaborative Filtering example using `keras` library.

## Preliminaries

### Imports

In [95]:
import os
import pickle

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split,KFold

import tensorflow as tf
from tensorflow import keras
import tensorflow.keras.backend as K
print(tf.__version__)

%matplotlib inline

1.11.0


### Tensorflow Config

In [2]:
#
from tensorflow.keras.backend import set_session

config = tf.ConfigProto()
config.gpu_options.allow_growth = True  # dynamically grow the memory used on the GPU
config.log_device_placement = True  # to log device placement (on which device the operation ran)
                                    # (nothing gets printed in Jupyter, only if you run it standalone)

sess = tf.Session(config=config)

set_session(sess)  # set this TensorFlow session as the default session for Keras

### Random Seed

In [3]:
seed=5543
np.random.seed(seed)

### Data

In [4]:
raw_dir="../raw/ml-1m"
data_dir="../data/MovieLens"

In [5]:
if not (os.path.exists(data_dir)):
    os.mkdir(data_dir)

In [6]:
filename=raw_dir+"/ratings.dat"

In [7]:
data_all=pd.read_csv(filename,sep="::",header=None,names=["userId","movieId","rating","TimeStamp"])

/home/manel/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:1: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.
  """Entry point for launching an IPython kernel.


In [8]:
data_all.head()

userId  movieId  rating  TimeStamp
0       1     1193       5  978300760
1       1      661       3  978302109
2       1      914       3  978301968
3       1     3408       4  978300275
4       1     2355       5  978824291

### Train Test Split

In this example we do not handle carefuly new users or new movies (we just asign them at random).

A production product would treat new users/movies separately.

In [9]:
userEncoder=LabelEncoder()
movieEncoder=LabelEncoder()

In [10]:
users_all=userEncoder.fit_transform(data_all[["userId"]].values.ravel())
movies_all=movieEncoder.fit_transform(data_all[["movieId"]].values.ravel())
ratings_all=data_all[["rating"]].values.ravel()

In [11]:
users,users_test,movies,movies_test,ratings, ratings_test=train_test_split(users_all,movies_all,ratings_all,test_size=0.15)
users_train,users_val, movies_train,movies_val, ratings_train,ratings_val=train_test_split(users,movies,ratings,test_size=0.15)
users_train.shape,users_val.shape,users_test.shape

((722650,), (127527,), (150032,))

In [12]:
unkown_users=~np.isin(users_val,users)
unknown_movies=~np.isin(movies_val,movies)
print(unkown_users.sum(),unknown_movies.sum())

0 0


In [13]:
unique_users=userEncoder.classes_
unique_movies=movieEncoder.classes_

In [14]:
N_users=len(unique_users)
N_movies=len(unique_movies)
print(N_users,N_movies)

6040 3706


In [15]:
mu=ratings_train.mean()
drating=np.mean((ratings_train-mu)**2)
print(mu,drating)

3.5812398809935653 1.2487443009293984


In [40]:
dratings_train=ratings_train-mu
dratings_val=ratings_val-mu
dratings_test=ratings_test-mu

## Keras Factorization Model

We define a metric for reports (keras by default will include the penalty on the metrics report)

In [67]:
def mse_err(y_true, y_pred):
    #return K.mean((y_pred-y_true)**2,axis=-1)
    return K.mean(K.square(y_pred - y_true), axis=-1)

We implement a call back to that we can see optimization progress every 10 epochs or so.

In [68]:
class ReportCallback(keras.callbacks.Callback):
    def __init__(self,frequency,use_val=False):
        self.freq=frequency
        self.use_val=use_val
        self.separator=" || "
        if not(self.use_val):
            self.separator="\n"
    def on_epoch_end(self, epoch, logs={}):
        if (epoch % self.freq ==0):
            train_loss=logs["loss"]
            train_error=logs["mse_err"]
            print(f"\t{epoch}: TRAIN Loss {train_loss:.4f},  Err {train_error:.4f}",end=self.separator)
            if self.use_val:
                val_loss=logs["val_loss"]
                val_error=logs["val_mse_err"]
                print(f"VAL Loss {val_loss:.4f}, Err {val_error:.4f}")

### Keras Rating Model

This implements
\begin{equation}
	\Delta \hat{r}_{i,u} =b_u + b_i + p_u^T q_i
\end{equation}

Where 
$$
    \Delta \hat{r}_{i,u} = \hat{r}_{i,u} - \mu
$$
using `keras` layers.

In [69]:
def collaborative_filter(U,M,F,penalty):
    users = keras.layers.Input(shape=(1,))
    movies = keras.layers.Input(shape=(1,))
    bu= keras.layers.Embedding(U,1,input_length=1,
                              embeddings_initializer=keras.initializers.RandomNormal(stddev=0.0001),
                              )(users)
    bm= keras.layers.Embedding(M,1,input_length=1,
                              embeddings_initializer=keras.initializers.RandomNormal(stddev=0.0001),
                              )(movies)
    pu= keras.layers.Embedding(U,F,input_length=1,
                               embeddings_initializer=keras.initializers.RandomNormal(stddev=1/np.sqrt(F)),
                               embeddings_regularizer=keras.regularizers.l2(penalty))(users)
    pm= keras.layers.Embedding(M,F,input_length=1,
                               embeddings_initializer=keras.initializers.RandomNormal(stddev=1/np.sqrt(F)),
                               embeddings_regularizer=keras.regularizers.l2(penalty))(movies)
    interaction=keras.layers.Dot(axes=-1)([pu,pm])
    agg=keras.layers.Add()([bu,bm,interaction])
    r_hat=keras.layers.Flatten()(agg)
    model = keras.models.Model(inputs=[users,movies], outputs=r_hat)
    return model

In [70]:
penalty=1e-5
F=10

We  can create a model for our choice of embedding dimention and penalty

In [71]:
model=collaborative_filter(N_users,N_movies,F,penalty)

We can compile the model and use it like any other `keras` model.

In [72]:
model.compile(optimizer=keras.optimizers.Adam(),
              metrics=[mse_err],
              loss='mean_squared_error')

In [73]:
model.summary()

__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to                     
input_33 (InputLayer)           (None, 1)            0                                            
__________________________________________________________________________________________________
input_34 (InputLayer)           (None, 1)            0                                            
__________________________________________________________________________________________________
embedding_66 (Embedding)        (None, 1, 10)        60400       input_33[0][0]                   
__________________________________________________________________________________________________
embedding_67 (Embedding)        (None, 1, 10)        37060       input_34[0][0]                   
__________________________________________________________________________________________________
embedding_

### Model Tranining

We can train as before, but now we must pass multiple inputs as a list of python objects.

In [74]:

model.fit([users_train,movies_train],dratings_train,
          epochs=200,batch_size=5000,
          callbacks=[ReportCallback(1,True)],
          verbose=0,
          validation_data=([users_val,movies_val],dratings_val))

/home/manel/anaconda3/lib/python3.6/site-packages/tensorflow/python/ops/gradients_impl.py:108: UserWarning: Converting sparse IndexedSlices to a dense Tensor of unknown shape. This may consume a large amount of memory.
  "Converting sparse IndexedSlices to a dense Tensor of unknown shape. "


	0: TRAIN Loss 1.3730,  Err 1.2827 || VAL Loss 1.3048, Err 1.2209
	1: TRAIN Loss 1.2413,  Err 1.1624 || VAL Loss 1.2022, Err 1.1282
	2: TRAIN Loss 1.1442,  Err 1.0744 || VAL Loss 1.1256, Err 1.0600
	3: TRAIN Loss 1.0713,  Err 1.0092 || VAL Loss 1.0680, Err 1.0093
	4: TRAIN Loss 1.0161,  Err 0.9604 || VAL Loss 1.0243, Err 0.9715
	5: TRAIN Loss 0.9737,  Err 0.9235 || VAL Loss 0.9907, Err 0.9429
	6: TRAIN Loss 0.9408,  Err 0.8951 || VAL Loss 0.9646, Err 0.9211
	7: TRAIN Loss 0.9147,  Err 0.8730 || VAL Loss 0.9441, Err 0.9041
	8: TRAIN Loss 0.8938,  Err 0.8554 || VAL Loss 0.9278, Err 0.8909
	9: TRAIN Loss 0.8769,  Err 0.8412 || VAL Loss 0.9146, Err 0.8803
	10: TRAIN Loss 0.8628,  Err 0.8295 || VAL Loss 0.9039, Err 0.8718
	11: TRAIN Loss 0.8509,  Err 0.8196 || VAL Loss 0.8951, Err 0.8647
	12: TRAIN Loss 0.8408,  Err 0.8111 || VAL Loss 0.8878, Err 0.8588
	13: TRAIN Loss 0.8319,  Err 0.8035 || VAL Loss 0.8816, Err 0.8537
	14: TRAIN Loss 0.8240,  Err 0.7965 || VAL Loss 0.8762, Err 0.8492
	15: 

	123: TRAIN Loss 0.6626,  Err 0.5957 || VAL Loss 0.8099, Err 0.7430
	124: TRAIN Loss 0.6625,  Err 0.5955 || VAL Loss 0.8098, Err 0.7428
	125: TRAIN Loss 0.6623,  Err 0.5953 || VAL Loss 0.8097, Err 0.7426
	126: TRAIN Loss 0.6622,  Err 0.5951 || VAL Loss 0.8097, Err 0.7425
	127: TRAIN Loss 0.6621,  Err 0.5949 || VAL Loss 0.8096, Err 0.7424
	128: TRAIN Loss 0.6620,  Err 0.5947 || VAL Loss 0.8095, Err 0.7423
	129: TRAIN Loss 0.6618,  Err 0.5945 || VAL Loss 0.8094, Err 0.7421
	130: TRAIN Loss 0.6617,  Err 0.5944 || VAL Loss 0.8093, Err 0.7419
	131: TRAIN Loss 0.6616,  Err 0.5942 || VAL Loss 0.8092, Err 0.7418
	132: TRAIN Loss 0.6615,  Err 0.5940 || VAL Loss 0.8091, Err 0.7416
	133: TRAIN Loss 0.6614,  Err 0.5938 || VAL Loss 0.8090, Err 0.7415
	134: TRAIN Loss 0.6613,  Err 0.5937 || VAL Loss 0.8089, Err 0.7413
	135: TRAIN Loss 0.6612,  Err 0.5935 || VAL Loss 0.8088, Err 0.7412
	136: TRAIN Loss 0.6610,  Err 0.5934 || VAL Loss 0.8088, Err 0.7411
	137: TRAIN Loss 0.6610,  Err 0.5932 || VAL Loss

KeyboardInterrupt: 

In [75]:
Y_pred=model.predict([users_val,movies_val]).ravel()
loss=np.mean((dratings_val-Y_pred)**2)
print(loss)

0.7354717149159153


## Hyperparameter search

In [79]:
results=[]
best_loss=1e10
best_F=None
best_penalty=None
steps=100
if True:
  for F in [1,5,10,15,20,30,50,100,150]: #[0,1,5,10,20,30,50,75,100,125,150]:
    for penalty in [0,1e-6,5e-6,1e-5,2e-5,3e-5,5e-5,1e-4]:
        print()
        print(f"F {F}, penalty {penalty} :")
        penalty=penalty
        model=collaborative_filter(N_users,N_movies,F,penalty)
        model.compile(optimizer=keras.optimizers.Adam(),
              metrics=[mse_err],
              loss='mean_squared_error')
        model.fit([users_train,movies_train],dratings_train,
          epochs=steps,batch_size=5000,
          callbacks=[ReportCallback(1,True)],
          verbose=0,
          validation_data=([users_val,movies_val],dratings_val))
        Y_pred=mu+model.predict([users_val,movies_val]).ravel()
        loss=np.mean((ratings_val-Y_pred)**2)
        results.append((F,penalty,loss))
        if loss<best_loss:
            best_loss=loss
            best_F=F
            best_penalty=penalty
        print()
        print(f"==> {F},{penalty},{loss} == best ({best_F},{best_penalty},{best_loss}) =============")


F 1, penalty 0 :


/home/manel/anaconda3/lib/python3.6/site-packages/tensorflow/python/ops/gradients_impl.py:108: UserWarning: Converting sparse IndexedSlices to a dense Tensor of unknown shape. This may consume a large amount of memory.
  "Converting sparse IndexedSlices to a dense Tensor of unknown shape. "


	0: TRAIN Loss 2.0850,  Err 2.0850 || VAL Loss 1.9501, Err 1.9501
	1: TRAIN Loss 1.8104,  Err 1.8104 || VAL Loss 1.7180, Err 1.7180
	2: TRAIN Loss 1.6053,  Err 1.6053 || VAL Loss 1.5429, Err 1.5429
	3: TRAIN Loss 1.4498,  Err 1.4498 || VAL Loss 1.4088, Err 1.4088
	4: TRAIN Loss 1.3303,  Err 1.3303 || VAL Loss 1.3053, Err 1.3053
	5: TRAIN Loss 1.2372,  Err 1.2372 || VAL Loss 1.2239, Err 1.2239
	6: TRAIN Loss 1.1638,  Err 1.1638 || VAL Loss 1.1595, Err 1.1595
	7: TRAIN Loss 1.1054,  Err 1.1054 || VAL Loss 1.1077, Err 1.1077
	8: TRAIN Loss 1.0583,  Err 1.0583 || VAL Loss 1.0658, Err 1.0658
	9: TRAIN Loss 1.0199,  Err 1.0199 || VAL Loss 1.0314, Err 1.0314
	10: TRAIN Loss 0.9883,  Err 0.9883 || VAL Loss 1.0031, Err 1.0031
	11: TRAIN Loss 0.9622,  Err 0.9622 || VAL Loss 0.9795, Err 0.9795
	12: TRAIN Loss 0.9404,  Err 0.9404 || VAL Loss 0.9597, Err 0.9597
	13: TRAIN Loss 0.9220,  Err 0.9220 || VAL Loss 0.9429, Err 0.9429
	14: TRAIN Loss 0.9064,  Err 0.9064 || VAL Loss 0.9287, Err 0.9287
	15: 

	22: TRAIN Loss 0.8422,  Err 0.8386 || VAL Loss 0.8644, Err 0.8608
	23: TRAIN Loss 0.8379,  Err 0.8344 || VAL Loss 0.8606, Err 0.8572
	24: TRAIN Loss 0.8342,  Err 0.8308 || VAL Loss 0.8573, Err 0.8539
	25: TRAIN Loss 0.8308,  Err 0.8275 || VAL Loss 0.8544, Err 0.8511
	26: TRAIN Loss 0.8279,  Err 0.8246 || VAL Loss 0.8518, Err 0.8486
	27: TRAIN Loss 0.8252,  Err 0.8220 || VAL Loss 0.8496, Err 0.8464
	28: TRAIN Loss 0.8229,  Err 0.8197 || VAL Loss 0.8475, Err 0.8444
	29: TRAIN Loss 0.8207,  Err 0.8177 || VAL Loss 0.8457, Err 0.8427
	30: TRAIN Loss 0.8188,  Err 0.8159 || VAL Loss 0.8441, Err 0.8411
	31: TRAIN Loss 0.8172,  Err 0.8142 || VAL Loss 0.8427, Err 0.8398
	32: TRAIN Loss 0.8156,  Err 0.8127 || VAL Loss 0.8414, Err 0.8386
	33: TRAIN Loss 0.8142,  Err 0.8114 || VAL Loss 0.8403, Err 0.8375
	34: TRAIN Loss 0.8130,  Err 0.8102 || VAL Loss 0.8393, Err 0.8365
	35: TRAIN Loss 0.8118,  Err 0.8091 || VAL Loss 0.8384, Err 0.8357
	36: TRAIN Loss 0.8107,  Err 0.8081 || VAL Loss 0.8375, Err 0.

	43: TRAIN Loss 0.8098,  Err 0.8028 || VAL Loss 0.8359, Err 0.8290
	44: TRAIN Loss 0.8090,  Err 0.8022 || VAL Loss 0.8353, Err 0.8287
	45: TRAIN Loss 0.8082,  Err 0.8017 || VAL Loss 0.8347, Err 0.8283
	46: TRAIN Loss 0.8075,  Err 0.8012 || VAL Loss 0.8341, Err 0.8279
	47: TRAIN Loss 0.8067,  Err 0.8006 || VAL Loss 0.8336, Err 0.8276
	48: TRAIN Loss 0.8060,  Err 0.8001 || VAL Loss 0.8331, Err 0.8273
	49: TRAIN Loss 0.8053,  Err 0.7996 || VAL Loss 0.8326, Err 0.8269
	50: TRAIN Loss 0.8046,  Err 0.7990 || VAL Loss 0.8321, Err 0.8266
	51: TRAIN Loss 0.8039,  Err 0.7985 || VAL Loss 0.8316, Err 0.8262
	52: TRAIN Loss 0.8032,  Err 0.7979 || VAL Loss 0.8310, Err 0.8258
	53: TRAIN Loss 0.8025,  Err 0.7972 || VAL Loss 0.8305, Err 0.8253
	54: TRAIN Loss 0.8017,  Err 0.7965 || VAL Loss 0.8299, Err 0.8248
	55: TRAIN Loss 0.8009,  Err 0.7958 || VAL Loss 0.8293, Err 0.8243
	56: TRAIN Loss 0.8000,  Err 0.7950 || VAL Loss 0.8286, Err 0.8237
	57: TRAIN Loss 0.7991,  Err 0.7942 || VAL Loss 0.8278, Err 0.

	64: TRAIN Loss 0.7910,  Err 0.7853 || VAL Loss 0.8185, Err 0.8127
	65: TRAIN Loss 0.7893,  Err 0.7834 || VAL Loss 0.8170, Err 0.8109
	66: TRAIN Loss 0.7876,  Err 0.7814 || VAL Loss 0.8153, Err 0.8090
	67: TRAIN Loss 0.7859,  Err 0.7794 || VAL Loss 0.8137, Err 0.8071
	68: TRAIN Loss 0.7842,  Err 0.7774 || VAL Loss 0.8122, Err 0.8053
	69: TRAIN Loss 0.7826,  Err 0.7755 || VAL Loss 0.8107, Err 0.8035
	70: TRAIN Loss 0.7810,  Err 0.7737 || VAL Loss 0.8093, Err 0.8019
	71: TRAIN Loss 0.7796,  Err 0.7720 || VAL Loss 0.8080, Err 0.8003
	72: TRAIN Loss 0.7783,  Err 0.7704 || VAL Loss 0.8068, Err 0.7988
	73: TRAIN Loss 0.7770,  Err 0.7689 || VAL Loss 0.8057, Err 0.7974
	74: TRAIN Loss 0.7759,  Err 0.7675 || VAL Loss 0.8047, Err 0.7961
	75: TRAIN Loss 0.7748,  Err 0.7662 || VAL Loss 0.8037, Err 0.7949
	76: TRAIN Loss 0.7738,  Err 0.7649 || VAL Loss 0.8029, Err 0.7939
	77: TRAIN Loss 0.7730,  Err 0.7638 || VAL Loss 0.8021, Err 0.7929
	78: TRAIN Loss 0.7721,  Err 0.7628 || VAL Loss 0.8014, Err 0.

	85: TRAIN Loss 0.7776,  Err 0.7617 || VAL Loss 0.8050, Err 0.7889
	86: TRAIN Loss 0.7774,  Err 0.7613 || VAL Loss 0.8049, Err 0.7887
	87: TRAIN Loss 0.7773,  Err 0.7610 || VAL Loss 0.8048, Err 0.7884
	88: TRAIN Loss 0.7771,  Err 0.7606 || VAL Loss 0.8048, Err 0.7882
	89: TRAIN Loss 0.7769,  Err 0.7603 || VAL Loss 0.8046, Err 0.7879
	90: TRAIN Loss 0.7768,  Err 0.7600 || VAL Loss 0.8045, Err 0.7877
	91: TRAIN Loss 0.7766,  Err 0.7597 || VAL Loss 0.8045, Err 0.7875
	92: TRAIN Loss 0.7765,  Err 0.7595 || VAL Loss 0.8044, Err 0.7873
	93: TRAIN Loss 0.7764,  Err 0.7592 || VAL Loss 0.8044, Err 0.7872
	94: TRAIN Loss 0.7763,  Err 0.7590 || VAL Loss 0.8044, Err 0.7870
	95: TRAIN Loss 0.7762,  Err 0.7588 || VAL Loss 0.8043, Err 0.7869
	96: TRAIN Loss 0.7761,  Err 0.7586 || VAL Loss 0.8042, Err 0.7867
	97: TRAIN Loss 0.7760,  Err 0.7584 || VAL Loss 0.8042, Err 0.7866
	98: TRAIN Loss 0.7759,  Err 0.7582 || VAL Loss 0.8042, Err 0.7864
	99: TRAIN Loss 0.7758,  Err 0.7580 || VAL Loss 0.8042, Err 0.

	5: TRAIN Loss 1.4466,  Err 1.1867 || VAL Loss 1.4055, Err 1.1591
	6: TRAIN Loss 1.3488,  Err 1.1146 || VAL Loss 1.3186, Err 1.0965
	7: TRAIN Loss 1.2692,  Err 1.0580 || VAL Loss 1.2476, Err 1.0471
	8: TRAIN Loss 1.2037,  Err 1.0130 || VAL Loss 1.1888, Err 1.0077
	9: TRAIN Loss 1.1492,  Err 0.9770 || VAL Loss 1.1397, Err 0.9761
	10: TRAIN Loss 1.1035,  Err 0.9478 || VAL Loss 1.0983, Err 0.9505
	11: TRAIN Loss 1.0649,  Err 0.9242 || VAL Loss 1.0633, Err 0.9296
	12: TRAIN Loss 1.0319,  Err 0.9047 || VAL Loss 1.0333, Err 0.9125
	13: TRAIN Loss 1.0036,  Err 0.8887 || VAL Loss 1.0075, Err 0.8984
	14: TRAIN Loss 0.9792,  Err 0.8754 || VAL Loss 0.9853, Err 0.8867
	15: TRAIN Loss 0.9581,  Err 0.8643 || VAL Loss 0.9659, Err 0.8769
	16: TRAIN Loss 0.9397,  Err 0.8551 || VAL Loss 0.9490, Err 0.8688
	17: TRAIN Loss 0.9236,  Err 0.8473 || VAL Loss 0.9343, Err 0.8620
	18: TRAIN Loss 0.9095,  Err 0.8408 || VAL Loss 0.9213, Err 0.8562
	19: TRAIN Loss 0.8971,  Err 0.8352 || VAL Loss 0.9100, Err 0.8514


	26: TRAIN Loss 0.8431,  Err 0.8130 || VAL Loss 0.8589, Err 0.8308
	27: TRAIN Loss 0.8381,  Err 0.8118 || VAL Loss 0.8543, Err 0.8298
	28: TRAIN Loss 0.8337,  Err 0.8108 || VAL Loss 0.8502, Err 0.8289
	29: TRAIN Loss 0.8298,  Err 0.8100 || VAL Loss 0.8467, Err 0.8282
	30: TRAIN Loss 0.8265,  Err 0.8092 || VAL Loss 0.8437, Err 0.8276
	31: TRAIN Loss 0.8236,  Err 0.8086 || VAL Loss 0.8410, Err 0.8270
	32: TRAIN Loss 0.8211,  Err 0.8081 || VAL Loss 0.8387, Err 0.8266
	33: TRAIN Loss 0.8189,  Err 0.8077 || VAL Loss 0.8367, Err 0.8263
	34: TRAIN Loss 0.8170,  Err 0.8073 || VAL Loss 0.8349, Err 0.8259
	35: TRAIN Loss 0.8154,  Err 0.8070 || VAL Loss 0.8334, Err 0.8257
	36: TRAIN Loss 0.8140,  Err 0.8068 || VAL Loss 0.8321, Err 0.8254
	37: TRAIN Loss 0.8127,  Err 0.8066 || VAL Loss 0.8309, Err 0.8252
	38: TRAIN Loss 0.8117,  Err 0.8064 || VAL Loss 0.8299, Err 0.8251
	39: TRAIN Loss 0.8108,  Err 0.8062 || VAL Loss 0.8291, Err 0.8249
	40: TRAIN Loss 0.8100,  Err 0.8061 || VAL Loss 0.8284, Err 0.

	47: TRAIN Loss 0.6979,  Err 0.6979 || VAL Loss 0.8024, Err 0.8024
	48: TRAIN Loss 0.6960,  Err 0.6960 || VAL Loss 0.8017, Err 0.8017
	49: TRAIN Loss 0.6943,  Err 0.6943 || VAL Loss 0.8011, Err 0.8011
	50: TRAIN Loss 0.6926,  Err 0.6926 || VAL Loss 0.8005, Err 0.8005
	51: TRAIN Loss 0.6909,  Err 0.6909 || VAL Loss 0.7999, Err 0.7999
	52: TRAIN Loss 0.6893,  Err 0.6893 || VAL Loss 0.7993, Err 0.7993
	53: TRAIN Loss 0.6878,  Err 0.6878 || VAL Loss 0.7989, Err 0.7989
	54: TRAIN Loss 0.6864,  Err 0.6864 || VAL Loss 0.7984, Err 0.7984
	55: TRAIN Loss 0.6850,  Err 0.6850 || VAL Loss 0.7978, Err 0.7978
	56: TRAIN Loss 0.6836,  Err 0.6836 || VAL Loss 0.7973, Err 0.7973
	57: TRAIN Loss 0.6824,  Err 0.6824 || VAL Loss 0.7969, Err 0.7969
	58: TRAIN Loss 0.6811,  Err 0.6811 || VAL Loss 0.7965, Err 0.7965
	59: TRAIN Loss 0.6799,  Err 0.6799 || VAL Loss 0.7961, Err 0.7961
	60: TRAIN Loss 0.6787,  Err 0.6787 || VAL Loss 0.7957, Err 0.7957
	61: TRAIN Loss 0.6776,  Err 0.6776 || VAL Loss 0.7953, Err 0.

	68: TRAIN Loss 0.6790,  Err 0.6709 || VAL Loss 0.7928, Err 0.7846
	69: TRAIN Loss 0.6781,  Err 0.6700 || VAL Loss 0.7924, Err 0.7842
	70: TRAIN Loss 0.6773,  Err 0.6691 || VAL Loss 0.7919, Err 0.7837
	71: TRAIN Loss 0.6764,  Err 0.6682 || VAL Loss 0.7915, Err 0.7833
	72: TRAIN Loss 0.6756,  Err 0.6673 || VAL Loss 0.7911, Err 0.7828
	73: TRAIN Loss 0.6748,  Err 0.6665 || VAL Loss 0.7907, Err 0.7824
	74: TRAIN Loss 0.6740,  Err 0.6656 || VAL Loss 0.7904, Err 0.7820
	75: TRAIN Loss 0.6733,  Err 0.6648 || VAL Loss 0.7900, Err 0.7816
	76: TRAIN Loss 0.6725,  Err 0.6641 || VAL Loss 0.7896, Err 0.7811
	77: TRAIN Loss 0.6718,  Err 0.6633 || VAL Loss 0.7893, Err 0.7808
	78: TRAIN Loss 0.6711,  Err 0.6626 || VAL Loss 0.7890, Err 0.7804
	79: TRAIN Loss 0.6704,  Err 0.6618 || VAL Loss 0.7886, Err 0.7800
	80: TRAIN Loss 0.6698,  Err 0.6611 || VAL Loss 0.7883, Err 0.7796
	81: TRAIN Loss 0.6691,  Err 0.6604 || VAL Loss 0.7880, Err 0.7793
	82: TRAIN Loss 0.6685,  Err 0.6598 || VAL Loss 0.7877, Err 0.

	89: TRAIN Loss 0.6902,  Err 0.6626 || VAL Loss 0.7903, Err 0.7627
	90: TRAIN Loss 0.6897,  Err 0.6621 || VAL Loss 0.7900, Err 0.7623
	91: TRAIN Loss 0.6894,  Err 0.6616 || VAL Loss 0.7897, Err 0.7619
	92: TRAIN Loss 0.6890,  Err 0.6611 || VAL Loss 0.7893, Err 0.7614
	93: TRAIN Loss 0.6886,  Err 0.6607 || VAL Loss 0.7890, Err 0.7610
	94: TRAIN Loss 0.6883,  Err 0.6602 || VAL Loss 0.7887, Err 0.7607
	95: TRAIN Loss 0.6879,  Err 0.6598 || VAL Loss 0.7884, Err 0.7603
	96: TRAIN Loss 0.6876,  Err 0.6594 || VAL Loss 0.7881, Err 0.7599
	97: TRAIN Loss 0.6872,  Err 0.6589 || VAL Loss 0.7878, Err 0.7595
	98: TRAIN Loss 0.6869,  Err 0.6586 || VAL Loss 0.7875, Err 0.7591
	99: TRAIN Loss 0.6866,  Err 0.6582 || VAL Loss 0.7872, Err 0.7588

==> 5,5e-06,0.7587746313494995 == best (5,5e-06,0.7587746313494995) =============

F 5, penalty 1e-05 :
	0: TRAIN Loss 1.4678,  Err 1.3766 || VAL Loss 1.3875, Err 1.3021
	1: TRAIN Loss 1.3151,  Err 1.2345 || VAL Loss 1.2650, Err 1.1891
	2: TRAIN Loss 1.2032,  Er

	9: TRAIN Loss 0.9068,  Err 0.8590 || VAL Loss 0.9252, Err 0.8805
	10: TRAIN Loss 0.8897,  Err 0.8476 || VAL Loss 0.9105, Err 0.8710
	11: TRAIN Loss 0.8756,  Err 0.8384 || VAL Loss 0.8984, Err 0.8634
	12: TRAIN Loss 0.8639,  Err 0.8308 || VAL Loss 0.8884, Err 0.8572
	13: TRAIN Loss 0.8542,  Err 0.8246 || VAL Loss 0.8800, Err 0.8520
	14: TRAIN Loss 0.8460,  Err 0.8194 || VAL Loss 0.8729, Err 0.8476
	15: TRAIN Loss 0.8390,  Err 0.8149 || VAL Loss 0.8669, Err 0.8439
	16: TRAIN Loss 0.8331,  Err 0.8110 || VAL Loss 0.8617, Err 0.8406
	17: TRAIN Loss 0.8279,  Err 0.8075 || VAL Loss 0.8573, Err 0.8377
	18: TRAIN Loss 0.8233,  Err 0.8042 || VAL Loss 0.8534, Err 0.8350
	19: TRAIN Loss 0.8191,  Err 0.8011 || VAL Loss 0.8498, Err 0.8323
	20: TRAIN Loss 0.8152,  Err 0.7980 || VAL Loss 0.8465, Err 0.8297
	21: TRAIN Loss 0.8115,  Err 0.7948 || VAL Loss 0.8434, Err 0.8269
	22: TRAIN Loss 0.8078,  Err 0.7914 || VAL Loss 0.8404, Err 0.8240
	23: TRAIN Loss 0.8041,  Err 0.7877 || VAL Loss 0.8374, Err 0.8

	30: TRAIN Loss 0.7887,  Err 0.7684 || VAL Loss 0.8194, Err 0.7983
	31: TRAIN Loss 0.7864,  Err 0.7645 || VAL Loss 0.8180, Err 0.7953
	32: TRAIN Loss 0.7843,  Err 0.7608 || VAL Loss 0.8168, Err 0.7926
	33: TRAIN Loss 0.7824,  Err 0.7574 || VAL Loss 0.8158, Err 0.7900
	34: TRAIN Loss 0.7807,  Err 0.7541 || VAL Loss 0.8149, Err 0.7877
	35: TRAIN Loss 0.7791,  Err 0.7511 || VAL Loss 0.8142, Err 0.7855
	36: TRAIN Loss 0.7777,  Err 0.7482 || VAL Loss 0.8136, Err 0.7835
	37: TRAIN Loss 0.7763,  Err 0.7455 || VAL Loss 0.8131, Err 0.7816
	38: TRAIN Loss 0.7751,  Err 0.7430 || VAL Loss 0.8126, Err 0.7799
	39: TRAIN Loss 0.7740,  Err 0.7406 || VAL Loss 0.8123, Err 0.7783
	40: TRAIN Loss 0.7729,  Err 0.7384 || VAL Loss 0.8119, Err 0.7768
	41: TRAIN Loss 0.7720,  Err 0.7363 || VAL Loss 0.8116, Err 0.7754
	42: TRAIN Loss 0.7711,  Err 0.7343 || VAL Loss 0.8114, Err 0.7741
	43: TRAIN Loss 0.7703,  Err 0.7324 || VAL Loss 0.8112, Err 0.7729
	44: TRAIN Loss 0.7695,  Err 0.7306 || VAL Loss 0.8110, Err 0.

	51: TRAIN Loss 0.7897,  Err 0.7545 || VAL Loss 0.8196, Err 0.7843
	52: TRAIN Loss 0.7895,  Err 0.7538 || VAL Loss 0.8197, Err 0.7838
	53: TRAIN Loss 0.7893,  Err 0.7531 || VAL Loss 0.8198, Err 0.7833
	54: TRAIN Loss 0.7892,  Err 0.7525 || VAL Loss 0.8198, Err 0.7829
	55: TRAIN Loss 0.7890,  Err 0.7518 || VAL Loss 0.8199, Err 0.7825
	56: TRAIN Loss 0.7889,  Err 0.7512 || VAL Loss 0.8200, Err 0.7821
	57: TRAIN Loss 0.7887,  Err 0.7506 || VAL Loss 0.8200, Err 0.7817
	58: TRAIN Loss 0.7886,  Err 0.7500 || VAL Loss 0.8201, Err 0.7813
	59: TRAIN Loss 0.7885,  Err 0.7494 || VAL Loss 0.8202, Err 0.7810
	60: TRAIN Loss 0.7884,  Err 0.7489 || VAL Loss 0.8203, Err 0.7806
	61: TRAIN Loss 0.7882,  Err 0.7483 || VAL Loss 0.8204, Err 0.7803
	62: TRAIN Loss 0.7881,  Err 0.7478 || VAL Loss 0.8205, Err 0.7800
	63: TRAIN Loss 0.7880,  Err 0.7472 || VAL Loss 0.8205, Err 0.7796
	64: TRAIN Loss 0.7879,  Err 0.7467 || VAL Loss 0.8206, Err 0.7793
	65: TRAIN Loss 0.7878,  Err 0.7463 || VAL Loss 0.8207, Err 0.

	72: TRAIN Loss 0.8050,  Err 0.8007 || VAL Loss 0.8246, Err 0.8203
	73: TRAIN Loss 0.8049,  Err 0.8006 || VAL Loss 0.8246, Err 0.8202
	74: TRAIN Loss 0.8049,  Err 0.8006 || VAL Loss 0.8245, Err 0.8202
	75: TRAIN Loss 0.8049,  Err 0.8005 || VAL Loss 0.8246, Err 0.8202
	76: TRAIN Loss 0.8049,  Err 0.8005 || VAL Loss 0.8246, Err 0.8202
	77: TRAIN Loss 0.8049,  Err 0.8004 || VAL Loss 0.8246, Err 0.8201
	78: TRAIN Loss 0.8049,  Err 0.8004 || VAL Loss 0.8246, Err 0.8201
	79: TRAIN Loss 0.8049,  Err 0.8003 || VAL Loss 0.8246, Err 0.8200
	80: TRAIN Loss 0.8049,  Err 0.8003 || VAL Loss 0.8246, Err 0.8200
	81: TRAIN Loss 0.8049,  Err 0.8002 || VAL Loss 0.8246, Err 0.8200
	82: TRAIN Loss 0.8049,  Err 0.8002 || VAL Loss 0.8246, Err 0.8200
	83: TRAIN Loss 0.8049,  Err 0.8002 || VAL Loss 0.8246, Err 0.8199
	84: TRAIN Loss 0.8049,  Err 0.8001 || VAL Loss 0.8246, Err 0.8199
	85: TRAIN Loss 0.8049,  Err 0.8001 || VAL Loss 0.8246, Err 0.8199
	86: TRAIN Loss 0.8048,  Err 0.8001 || VAL Loss 0.8247, Err 0.

	93: TRAIN Loss 0.5781,  Err 0.5781 || VAL Loss 0.8233, Err 0.8233
	94: TRAIN Loss 0.5776,  Err 0.5776 || VAL Loss 0.8235, Err 0.8235
	95: TRAIN Loss 0.5770,  Err 0.5770 || VAL Loss 0.8237, Err 0.8237
	96: TRAIN Loss 0.5765,  Err 0.5765 || VAL Loss 0.8238, Err 0.8238
	97: TRAIN Loss 0.5760,  Err 0.5760 || VAL Loss 0.8239, Err 0.8239
	98: TRAIN Loss 0.5754,  Err 0.5754 || VAL Loss 0.8242, Err 0.8242
	99: TRAIN Loss 0.5749,  Err 0.5749 || VAL Loss 0.8243, Err 0.8243

==> 10,0,0.8242818950603251 == best (5,1e-05,0.7424955281761261) =============

F 10, penalty 1e-06 :
	0: TRAIN Loss 1.2962,  Err 1.2868 || VAL Loss 1.2379, Err 1.2288
	1: TRAIN Loss 1.1778,  Err 1.1690 || VAL Loss 1.1469, Err 1.1384
	2: TRAIN Loss 1.0903,  Err 1.0820 || VAL Loss 1.0794, Err 1.0714
	3: TRAIN Loss 1.0247,  Err 1.0168 || VAL Loss 1.0290, Err 1.0213
	4: TRAIN Loss 0.9750,  Err 0.9675 || VAL Loss 0.9912, Err 0.9838
	5: TRAIN Loss 0.9368,  Err 0.9296 || VAL Loss 0.9624, Err 0.9552
	6: TRAIN Loss 0.9071,  Err 0.90

	13: TRAIN Loss 0.8191,  Err 0.7976 || VAL Loss 0.8783, Err 0.8569
	14: TRAIN Loss 0.8104,  Err 0.7891 || VAL Loss 0.8729, Err 0.8518
	15: TRAIN Loss 0.8021,  Err 0.7810 || VAL Loss 0.8680, Err 0.8469
	16: TRAIN Loss 0.7941,  Err 0.7730 || VAL Loss 0.8633, Err 0.8422
	17: TRAIN Loss 0.7861,  Err 0.7649 || VAL Loss 0.8588, Err 0.8375
	18: TRAIN Loss 0.7783,  Err 0.7569 || VAL Loss 0.8544, Err 0.8328
	19: TRAIN Loss 0.7705,  Err 0.7488 || VAL Loss 0.8502, Err 0.8283
	20: TRAIN Loss 0.7629,  Err 0.7408 || VAL Loss 0.8463, Err 0.8239
	21: TRAIN Loss 0.7556,  Err 0.7330 || VAL Loss 0.8426, Err 0.8197
	22: TRAIN Loss 0.7486,  Err 0.7254 || VAL Loss 0.8394, Err 0.8159
	23: TRAIN Loss 0.7420,  Err 0.7181 || VAL Loss 0.8365, Err 0.8123
	24: TRAIN Loss 0.7358,  Err 0.7113 || VAL Loss 0.8340, Err 0.8092
	25: TRAIN Loss 0.7301,  Err 0.7049 || VAL Loss 0.8318, Err 0.8064
	26: TRAIN Loss 0.7247,  Err 0.6989 || VAL Loss 0.8300, Err 0.8039
	27: TRAIN Loss 0.7198,  Err 0.6934 || VAL Loss 0.8285, Err 0.

	34: TRAIN Loss 0.7150,  Err 0.6725 || VAL Loss 0.8198, Err 0.7770
	35: TRAIN Loss 0.7125,  Err 0.6692 || VAL Loss 0.8194, Err 0.7756
	36: TRAIN Loss 0.7103,  Err 0.6661 || VAL Loss 0.8190, Err 0.7744
	37: TRAIN Loss 0.7082,  Err 0.6631 || VAL Loss 0.8186, Err 0.7732
	38: TRAIN Loss 0.7062,  Err 0.6603 || VAL Loss 0.8183, Err 0.7720
	39: TRAIN Loss 0.7043,  Err 0.6577 || VAL Loss 0.8179, Err 0.7710
	40: TRAIN Loss 0.7025,  Err 0.6551 || VAL Loss 0.8177, Err 0.7700
	41: TRAIN Loss 0.7008,  Err 0.6527 || VAL Loss 0.8174, Err 0.7690
	42: TRAIN Loss 0.6992,  Err 0.6504 || VAL Loss 0.8171, Err 0.7681
	43: TRAIN Loss 0.6977,  Err 0.6482 || VAL Loss 0.8169, Err 0.7671
	44: TRAIN Loss 0.6962,  Err 0.6461 || VAL Loss 0.8167, Err 0.7664
	45: TRAIN Loss 0.6949,  Err 0.6442 || VAL Loss 0.8165, Err 0.7655
	46: TRAIN Loss 0.6935,  Err 0.6423 || VAL Loss 0.8163, Err 0.7647
	47: TRAIN Loss 0.6923,  Err 0.6405 || VAL Loss 0.8161, Err 0.7640
	48: TRAIN Loss 0.6911,  Err 0.6387 || VAL Loss 0.8159, Err 0.

	55: TRAIN Loss 0.7236,  Err 0.6532 || VAL Loss 0.8177, Err 0.7470
	56: TRAIN Loss 0.7231,  Err 0.6521 || VAL Loss 0.8178, Err 0.7466
	57: TRAIN Loss 0.7227,  Err 0.6512 || VAL Loss 0.8178, Err 0.7461
	58: TRAIN Loss 0.7222,  Err 0.6503 || VAL Loss 0.8178, Err 0.7457
	59: TRAIN Loss 0.7219,  Err 0.6494 || VAL Loss 0.8179, Err 0.7453
	60: TRAIN Loss 0.7215,  Err 0.6486 || VAL Loss 0.8179, Err 0.7448
	61: TRAIN Loss 0.7211,  Err 0.6478 || VAL Loss 0.8180, Err 0.7445
	62: TRAIN Loss 0.7207,  Err 0.6471 || VAL Loss 0.8180, Err 0.7441
	63: TRAIN Loss 0.7204,  Err 0.6463 || VAL Loss 0.8180, Err 0.7438
	64: TRAIN Loss 0.7201,  Err 0.6456 || VAL Loss 0.8181, Err 0.7435
	65: TRAIN Loss 0.7198,  Err 0.6450 || VAL Loss 0.8181, Err 0.7432
	66: TRAIN Loss 0.7195,  Err 0.6444 || VAL Loss 0.8181, Err 0.7428
	67: TRAIN Loss 0.7192,  Err 0.6437 || VAL Loss 0.8181, Err 0.7425
	68: TRAIN Loss 0.7190,  Err 0.6432 || VAL Loss 0.8182, Err 0.7423
	69: TRAIN Loss 0.7187,  Err 0.6426 || VAL Loss 0.8181, Err 0.

	76: TRAIN Loss 0.7484,  Err 0.6690 || VAL Loss 0.8237, Err 0.7442
	77: TRAIN Loss 0.7484,  Err 0.6687 || VAL Loss 0.8238, Err 0.7441
	78: TRAIN Loss 0.7483,  Err 0.6684 || VAL Loss 0.8239, Err 0.7440
	79: TRAIN Loss 0.7482,  Err 0.6681 || VAL Loss 0.8240, Err 0.7438
	80: TRAIN Loss 0.7481,  Err 0.6678 || VAL Loss 0.8241, Err 0.7438
	81: TRAIN Loss 0.7480,  Err 0.6676 || VAL Loss 0.8242, Err 0.7437
	82: TRAIN Loss 0.7479,  Err 0.6673 || VAL Loss 0.8241, Err 0.7435
	83: TRAIN Loss 0.7479,  Err 0.6671 || VAL Loss 0.8243, Err 0.7434
	84: TRAIN Loss 0.7478,  Err 0.6668 || VAL Loss 0.8243, Err 0.7433
	85: TRAIN Loss 0.7477,  Err 0.6666 || VAL Loss 0.8244, Err 0.7432
	86: TRAIN Loss 0.7477,  Err 0.6664 || VAL Loss 0.8244, Err 0.7431
	87: TRAIN Loss 0.7476,  Err 0.6662 || VAL Loss 0.8245, Err 0.7430
	88: TRAIN Loss 0.7475,  Err 0.6660 || VAL Loss 0.8245, Err 0.7429
	89: TRAIN Loss 0.7475,  Err 0.6658 || VAL Loss 0.8245, Err 0.7428
	90: TRAIN Loss 0.7474,  Err 0.6656 || VAL Loss 0.8246, Err 0.

	97: TRAIN Loss 0.7854,  Err 0.7261 || VAL Loss 0.8278, Err 0.7685
	98: TRAIN Loss 0.7854,  Err 0.7261 || VAL Loss 0.8278, Err 0.7685
	99: TRAIN Loss 0.7854,  Err 0.7260 || VAL Loss 0.8279, Err 0.7685

==> 10,5e-05,0.7684873457920898 == best (10,2e-05,0.7366513533519797) =============

F 10, penalty 0.0001 :
	0: TRAIN Loss 2.0548,  Err 1.2688 || VAL Loss 1.8180, Err 1.1984
	1: TRAIN Loss 1.6371,  Err 1.1372 || VAL Loss 1.4915, Err 1.0973
	2: TRAIN Loss 1.3676,  Err 1.0499 || VAL Loss 1.2786, Err 1.0284
	3: TRAIN Loss 1.1917,  Err 0.9901 || VAL Loss 1.1388, Err 0.9800
	4: TRAIN Loss 1.0760,  Err 0.9481 || VAL Loss 1.0461, Err 0.9453
	5: TRAIN Loss 0.9992,  Err 0.9179 || VAL Loss 0.9840, Err 0.9198
	6: TRAIN Loss 0.9475,  Err 0.8956 || VAL Loss 0.9418, Err 0.9007
	7: TRAIN Loss 0.9122,  Err 0.8789 || VAL Loss 0.9127, Err 0.8863
	8: TRAIN Loss 0.8876,  Err 0.8661 || VAL Loss 0.8922, Err 0.8751
	9: TRAIN Loss 0.8700,  Err 0.8561 || VAL Loss 0.8774, Err 0.8663
	10: TRAIN Loss 0.8573,  Err 0

	17: TRAIN Loss 0.7099,  Err 0.7099 || VAL Loss 0.8408, Err 0.8408
	18: TRAIN Loss 0.6999,  Err 0.6999 || VAL Loss 0.8384, Err 0.8384
	19: TRAIN Loss 0.6906,  Err 0.6906 || VAL Loss 0.8366, Err 0.8366
	20: TRAIN Loss 0.6819,  Err 0.6819 || VAL Loss 0.8353, Err 0.8353
	21: TRAIN Loss 0.6738,  Err 0.6738 || VAL Loss 0.8344, Err 0.8344
	22: TRAIN Loss 0.6662,  Err 0.6662 || VAL Loss 0.8339, Err 0.8339
	23: TRAIN Loss 0.6590,  Err 0.6590 || VAL Loss 0.8337, Err 0.8337
	24: TRAIN Loss 0.6523,  Err 0.6523 || VAL Loss 0.8338, Err 0.8338
	25: TRAIN Loss 0.6460,  Err 0.6460 || VAL Loss 0.8339, Err 0.8339
	26: TRAIN Loss 0.6401,  Err 0.6401 || VAL Loss 0.8343, Err 0.8343
	27: TRAIN Loss 0.6345,  Err 0.6345 || VAL Loss 0.8348, Err 0.8348
	28: TRAIN Loss 0.6293,  Err 0.6293 || VAL Loss 0.8353, Err 0.8353
	29: TRAIN Loss 0.6243,  Err 0.6243 || VAL Loss 0.8360, Err 0.8360
	30: TRAIN Loss 0.6196,  Err 0.6196 || VAL Loss 0.8366, Err 0.8366
	31: TRAIN Loss 0.6151,  Err 0.6151 || VAL Loss 0.8374, Err 0.

	38: TRAIN Loss 0.6041,  Err 0.5895 || VAL Loss 0.8441, Err 0.8294
	39: TRAIN Loss 0.6013,  Err 0.5864 || VAL Loss 0.8446, Err 0.8297
	40: TRAIN Loss 0.5986,  Err 0.5835 || VAL Loss 0.8451, Err 0.8300
	41: TRAIN Loss 0.5960,  Err 0.5808 || VAL Loss 0.8455, Err 0.8302
	42: TRAIN Loss 0.5935,  Err 0.5781 || VAL Loss 0.8461, Err 0.8306
	43: TRAIN Loss 0.5911,  Err 0.5755 || VAL Loss 0.8465, Err 0.8308
	44: TRAIN Loss 0.5889,  Err 0.5731 || VAL Loss 0.8470, Err 0.8311
	45: TRAIN Loss 0.5867,  Err 0.5708 || VAL Loss 0.8474, Err 0.8314
	46: TRAIN Loss 0.5846,  Err 0.5685 || VAL Loss 0.8478, Err 0.8316
	47: TRAIN Loss 0.5827,  Err 0.5664 || VAL Loss 0.8482, Err 0.8319
	48: TRAIN Loss 0.5808,  Err 0.5643 || VAL Loss 0.8487, Err 0.8321
	49: TRAIN Loss 0.5789,  Err 0.5623 || VAL Loss 0.8490, Err 0.8324
	50: TRAIN Loss 0.5772,  Err 0.5604 || VAL Loss 0.8494, Err 0.8326
	51: TRAIN Loss 0.5755,  Err 0.5586 || VAL Loss 0.8498, Err 0.8328
	52: TRAIN Loss 0.5739,  Err 0.5568 || VAL Loss 0.8502, Err 0.

	59: TRAIN Loss 0.6094,  Err 0.5540 || VAL Loss 0.8418, Err 0.7863
	60: TRAIN Loss 0.6086,  Err 0.5529 || VAL Loss 0.8419, Err 0.7862
	61: TRAIN Loss 0.6078,  Err 0.5518 || VAL Loss 0.8421, Err 0.7860
	62: TRAIN Loss 0.6070,  Err 0.5508 || VAL Loss 0.8422, Err 0.7859
	63: TRAIN Loss 0.6062,  Err 0.5498 || VAL Loss 0.8424, Err 0.7858
	64: TRAIN Loss 0.6055,  Err 0.5488 || VAL Loss 0.8425, Err 0.7857
	65: TRAIN Loss 0.6048,  Err 0.5479 || VAL Loss 0.8426, Err 0.7856
	66: TRAIN Loss 0.6042,  Err 0.5470 || VAL Loss 0.8427, Err 0.7855
	67: TRAIN Loss 0.6035,  Err 0.5462 || VAL Loss 0.8429, Err 0.7854
	68: TRAIN Loss 0.6029,  Err 0.5453 || VAL Loss 0.8430, Err 0.7854
	69: TRAIN Loss 0.6023,  Err 0.5445 || VAL Loss 0.8432, Err 0.7853
	70: TRAIN Loss 0.6017,  Err 0.5437 || VAL Loss 0.8433, Err 0.7852
	71: TRAIN Loss 0.6011,  Err 0.5430 || VAL Loss 0.8433, Err 0.7851
	72: TRAIN Loss 0.6006,  Err 0.5422 || VAL Loss 0.8435, Err 0.7850
	73: TRAIN Loss 0.6000,  Err 0.5415 || VAL Loss 0.8436, Err 0.

	80: TRAIN Loss 0.6424,  Err 0.5602 || VAL Loss 0.8381, Err 0.7559
	81: TRAIN Loss 0.6421,  Err 0.5598 || VAL Loss 0.8382, Err 0.7558
	82: TRAIN Loss 0.6419,  Err 0.5593 || VAL Loss 0.8383, Err 0.7557
	83: TRAIN Loss 0.6416,  Err 0.5589 || VAL Loss 0.8384, Err 0.7556
	84: TRAIN Loss 0.6413,  Err 0.5584 || VAL Loss 0.8384, Err 0.7555
	85: TRAIN Loss 0.6411,  Err 0.5581 || VAL Loss 0.8385, Err 0.7554
	86: TRAIN Loss 0.6408,  Err 0.5576 || VAL Loss 0.8385, Err 0.7553
	87: TRAIN Loss 0.6406,  Err 0.5572 || VAL Loss 0.8386, Err 0.7553
	88: TRAIN Loss 0.6403,  Err 0.5568 || VAL Loss 0.8388, Err 0.7552
	89: TRAIN Loss 0.6401,  Err 0.5565 || VAL Loss 0.8387, Err 0.7551
	90: TRAIN Loss 0.6399,  Err 0.5561 || VAL Loss 0.8388, Err 0.7550
	91: TRAIN Loss 0.6396,  Err 0.5558 || VAL Loss 0.8389, Err 0.7549
	92: TRAIN Loss 0.6394,  Err 0.5554 || VAL Loss 0.8389, Err 0.7548
	93: TRAIN Loss 0.6392,  Err 0.5551 || VAL Loss 0.8390, Err 0.7548
	94: TRAIN Loss 0.6390,  Err 0.5547 || VAL Loss 0.8390, Err 0.

	0: TRAIN Loss 1.5013,  Err 1.2496 || VAL Loss 1.4072, Err 1.1899
	1: TRAIN Loss 1.3223,  Err 1.1313 || VAL Loss 1.2669, Err 1.1004
	2: TRAIN Loss 1.1943,  Err 1.0470 || VAL Loss 1.1646, Err 1.0354
	3: TRAIN Loss 1.1010,  Err 0.9860 || VAL Loss 1.0893, Err 0.9876
	4: TRAIN Loss 1.0324,  Err 0.9414 || VAL Loss 1.0333, Err 0.9522
	5: TRAIN Loss 0.9815,  Err 0.9083 || VAL Loss 0.9914, Err 0.9256
	6: TRAIN Loss 0.9433,  Err 0.8834 || VAL Loss 0.9596, Err 0.9052
	7: TRAIN Loss 0.9143,  Err 0.8643 || VAL Loss 0.9352, Err 0.8893
	8: TRAIN Loss 0.8919,  Err 0.8493 || VAL Loss 0.9162, Err 0.8767
	9: TRAIN Loss 0.8743,  Err 0.8372 || VAL Loss 0.9012, Err 0.8663
	10: TRAIN Loss 0.8602,  Err 0.8270 || VAL Loss 0.8890, Err 0.8574
	11: TRAIN Loss 0.8485,  Err 0.8180 || VAL Loss 0.8789, Err 0.8495
	12: TRAIN Loss 0.8385,  Err 0.8097 || VAL Loss 0.8702, Err 0.8420
	13: TRAIN Loss 0.8296,  Err 0.8016 || VAL Loss 0.8626, Err 0.8347
	14: TRAIN Loss 0.8215,  Err 0.7934 || VAL Loss 0.8558, Err 0.8274
	15: 

	21: TRAIN Loss 0.8026,  Err 0.7745 || VAL Loss 0.8268, Err 0.7980
	22: TRAIN Loss 0.8008,  Err 0.7710 || VAL Loss 0.8258, Err 0.7953
	23: TRAIN Loss 0.7993,  Err 0.7680 || VAL Loss 0.8250, Err 0.7930
	24: TRAIN Loss 0.7980,  Err 0.7650 || VAL Loss 0.8244, Err 0.7909
	25: TRAIN Loss 0.7969,  Err 0.7625 || VAL Loss 0.8239, Err 0.7890
	26: TRAIN Loss 0.7959,  Err 0.7601 || VAL Loss 0.8236, Err 0.7874
	27: TRAIN Loss 0.7950,  Err 0.7579 || VAL Loss 0.8233, Err 0.7858
	28: TRAIN Loss 0.7942,  Err 0.7561 || VAL Loss 0.8231, Err 0.7846
	29: TRAIN Loss 0.7935,  Err 0.7543 || VAL Loss 0.8230, Err 0.7834
	30: TRAIN Loss 0.7929,  Err 0.7527 || VAL Loss 0.8229, Err 0.7824
	31: TRAIN Loss 0.7924,  Err 0.7512 || VAL Loss 0.8228, Err 0.7814
	32: TRAIN Loss 0.7918,  Err 0.7498 || VAL Loss 0.8228, Err 0.7805
	33: TRAIN Loss 0.7914,  Err 0.7486 || VAL Loss 0.8228, Err 0.7798
	34: TRAIN Loss 0.7910,  Err 0.7475 || VAL Loss 0.8228, Err 0.7790
	35: TRAIN Loss 0.7906,  Err 0.7464 || VAL Loss 0.8228, Err 0.

	42: TRAIN Loss 0.8062,  Err 0.8020 || VAL Loss 0.8250, Err 0.8208
	43: TRAIN Loss 0.8061,  Err 0.8019 || VAL Loss 0.8249, Err 0.8207
	44: TRAIN Loss 0.8060,  Err 0.8018 || VAL Loss 0.8248, Err 0.8207
	45: TRAIN Loss 0.8059,  Err 0.8017 || VAL Loss 0.8248, Err 0.8207
	46: TRAIN Loss 0.8059,  Err 0.8016 || VAL Loss 0.8248, Err 0.8206
	47: TRAIN Loss 0.8058,  Err 0.8016 || VAL Loss 0.8247, Err 0.8206
	48: TRAIN Loss 0.8057,  Err 0.8015 || VAL Loss 0.8247, Err 0.8206
	49: TRAIN Loss 0.8057,  Err 0.8014 || VAL Loss 0.8247, Err 0.8205
	50: TRAIN Loss 0.8056,  Err 0.8014 || VAL Loss 0.8247, Err 0.8205
	51: TRAIN Loss 0.8056,  Err 0.8013 || VAL Loss 0.8247, Err 0.8205
	52: TRAIN Loss 0.8056,  Err 0.8013 || VAL Loss 0.8247, Err 0.8204
	53: TRAIN Loss 0.8055,  Err 0.8013 || VAL Loss 0.8247, Err 0.8205
	54: TRAIN Loss 0.8055,  Err 0.8012 || VAL Loss 0.8246, Err 0.8204
	55: TRAIN Loss 0.8054,  Err 0.8012 || VAL Loss 0.8246, Err 0.8204
	56: TRAIN Loss 0.8054,  Err 0.8011 || VAL Loss 0.8246, Err 0.

	63: TRAIN Loss 0.4885,  Err 0.4885 || VAL Loss 0.8912, Err 0.8912
	64: TRAIN Loss 0.4873,  Err 0.4873 || VAL Loss 0.8924, Err 0.8924
	65: TRAIN Loss 0.4860,  Err 0.4860 || VAL Loss 0.8935, Err 0.8935
	66: TRAIN Loss 0.4849,  Err 0.4849 || VAL Loss 0.8947, Err 0.8947
	67: TRAIN Loss 0.4837,  Err 0.4837 || VAL Loss 0.8958, Err 0.8958
	68: TRAIN Loss 0.4826,  Err 0.4826 || VAL Loss 0.8970, Err 0.8970
	69: TRAIN Loss 0.4816,  Err 0.4816 || VAL Loss 0.8980, Err 0.8980
	70: TRAIN Loss 0.4805,  Err 0.4805 || VAL Loss 0.8992, Err 0.8992
	71: TRAIN Loss 0.4795,  Err 0.4795 || VAL Loss 0.9003, Err 0.9003
	72: TRAIN Loss 0.4785,  Err 0.4785 || VAL Loss 0.9015, Err 0.9015
	73: TRAIN Loss 0.4775,  Err 0.4775 || VAL Loss 0.9026, Err 0.9026
	74: TRAIN Loss 0.4766,  Err 0.4766 || VAL Loss 0.9038, Err 0.9038
	75: TRAIN Loss 0.4756,  Err 0.4756 || VAL Loss 0.9049, Err 0.9049
	76: TRAIN Loss 0.4747,  Err 0.4747 || VAL Loss 0.9060, Err 0.9060
	77: TRAIN Loss 0.4739,  Err 0.4739 || VAL Loss 0.9071, Err 0.

	84: TRAIN Loss 0.4940,  Err 0.4691 || VAL Loss 0.8977, Err 0.8727
	85: TRAIN Loss 0.4935,  Err 0.4684 || VAL Loss 0.8984, Err 0.8733
	86: TRAIN Loss 0.4929,  Err 0.4678 || VAL Loss 0.8990, Err 0.8738
	87: TRAIN Loss 0.4923,  Err 0.4671 || VAL Loss 0.8996, Err 0.8744
	88: TRAIN Loss 0.4918,  Err 0.4665 || VAL Loss 0.9003, Err 0.8750
	89: TRAIN Loss 0.4912,  Err 0.4659 || VAL Loss 0.9010, Err 0.8756
	90: TRAIN Loss 0.4907,  Err 0.4653 || VAL Loss 0.9016, Err 0.8761
	91: TRAIN Loss 0.4902,  Err 0.4647 || VAL Loss 0.9022, Err 0.8766
	92: TRAIN Loss 0.4897,  Err 0.4641 || VAL Loss 0.9028, Err 0.8772
	93: TRAIN Loss 0.4892,  Err 0.4635 || VAL Loss 0.9034, Err 0.8777
	94: TRAIN Loss 0.4887,  Err 0.4630 || VAL Loss 0.9039, Err 0.8782
	95: TRAIN Loss 0.4882,  Err 0.4624 || VAL Loss 0.9046, Err 0.8787
	96: TRAIN Loss 0.4878,  Err 0.4619 || VAL Loss 0.9051, Err 0.8792
	97: TRAIN Loss 0.4873,  Err 0.4614 || VAL Loss 0.9057, Err 0.8797
	98: TRAIN Loss 0.4869,  Err 0.4608 || VAL Loss 0.9063, Err 0.

	4: TRAIN Loss 0.9866,  Err 0.9326 || VAL Loss 1.0081, Err 0.9567
	5: TRAIN Loss 0.9450,  Err 0.8956 || VAL Loss 0.9769, Err 0.9295
	6: TRAIN Loss 0.9119,  Err 0.8661 || VAL Loss 0.9522, Err 0.9079
	7: TRAIN Loss 0.8847,  Err 0.8416 || VAL Loss 0.9322, Err 0.8901
	8: TRAIN Loss 0.8618,  Err 0.8205 || VAL Loss 0.9156, Err 0.8750
	9: TRAIN Loss 0.8419,  Err 0.8017 || VAL Loss 0.9017, Err 0.8618
	10: TRAIN Loss 0.8242,  Err 0.7844 || VAL Loss 0.8897, Err 0.8499
	11: TRAIN Loss 0.8082,  Err 0.7682 || VAL Loss 0.8795, Err 0.8393
	12: TRAIN Loss 0.7937,  Err 0.7531 || VAL Loss 0.8708, Err 0.8298
	13: TRAIN Loss 0.7805,  Err 0.7389 || VAL Loss 0.8637, Err 0.8215
	14: TRAIN Loss 0.7686,  Err 0.7256 || VAL Loss 0.8579, Err 0.8142
	15: TRAIN Loss 0.7578,  Err 0.7132 || VAL Loss 0.8532, Err 0.8078
	16: TRAIN Loss 0.7479,  Err 0.7016 || VAL Loss 0.8495, Err 0.8023
	17: TRAIN Loss 0.7390,  Err 0.6908 || VAL Loss 0.8468, Err 0.7977
	18: TRAIN Loss 0.7308,  Err 0.6807 || VAL Loss 0.8447, Err 0.7937
	

	25: TRAIN Loss 0.7346,  Err 0.6620 || VAL Loss 0.8349, Err 0.7613
	26: TRAIN Loss 0.7315,  Err 0.6567 || VAL Loss 0.8351, Err 0.7592
	27: TRAIN Loss 0.7287,  Err 0.6517 || VAL Loss 0.8355, Err 0.7574
	28: TRAIN Loss 0.7262,  Err 0.6470 || VAL Loss 0.8359, Err 0.7558
	29: TRAIN Loss 0.7239,  Err 0.6427 || VAL Loss 0.8364, Err 0.7544
	30: TRAIN Loss 0.7218,  Err 0.6387 || VAL Loss 0.8369, Err 0.7531
	31: TRAIN Loss 0.7198,  Err 0.6350 || VAL Loss 0.8376, Err 0.7520
	32: TRAIN Loss 0.7181,  Err 0.6316 || VAL Loss 0.8382, Err 0.7510
	33: TRAIN Loss 0.7164,  Err 0.6284 || VAL Loss 0.8388, Err 0.7501
	34: TRAIN Loss 0.7149,  Err 0.6253 || VAL Loss 0.8395, Err 0.7493
	35: TRAIN Loss 0.7135,  Err 0.6225 || VAL Loss 0.8402, Err 0.7485
	36: TRAIN Loss 0.7122,  Err 0.6198 || VAL Loss 0.8409, Err 0.7479
	37: TRAIN Loss 0.7110,  Err 0.6173 || VAL Loss 0.8416, Err 0.7473
	38: TRAIN Loss 0.7099,  Err 0.6150 || VAL Loss 0.8422, Err 0.7468
	39: TRAIN Loss 0.7088,  Err 0.6128 || VAL Loss 0.8429, Err 0.

	46: TRAIN Loss 0.7436,  Err 0.6465 || VAL Loss 0.8400, Err 0.7425
	47: TRAIN Loss 0.7433,  Err 0.6453 || VAL Loss 0.8404, Err 0.7422
	48: TRAIN Loss 0.7429,  Err 0.6442 || VAL Loss 0.8410, Err 0.7420
	49: TRAIN Loss 0.7426,  Err 0.6431 || VAL Loss 0.8414, Err 0.7417
	50: TRAIN Loss 0.7423,  Err 0.6422 || VAL Loss 0.8419, Err 0.7415
	51: TRAIN Loss 0.7421,  Err 0.6412 || VAL Loss 0.8424, Err 0.7414
	52: TRAIN Loss 0.7418,  Err 0.6403 || VAL Loss 0.8428, Err 0.7411
	53: TRAIN Loss 0.7416,  Err 0.6395 || VAL Loss 0.8432, Err 0.7410
	54: TRAIN Loss 0.7413,  Err 0.6387 || VAL Loss 0.8436, Err 0.7408
	55: TRAIN Loss 0.7411,  Err 0.6379 || VAL Loss 0.8440, Err 0.7407
	56: TRAIN Loss 0.7409,  Err 0.6371 || VAL Loss 0.8444, Err 0.7405
	57: TRAIN Loss 0.7407,  Err 0.6365 || VAL Loss 0.8448, Err 0.7404
	58: TRAIN Loss 0.7405,  Err 0.6359 || VAL Loss 0.8451, Err 0.7403
	59: TRAIN Loss 0.7403,  Err 0.6352 || VAL Loss 0.8454, Err 0.7401
	60: TRAIN Loss 0.7402,  Err 0.6345 || VAL Loss 0.8457, Err 0.

	67: TRAIN Loss 0.7865,  Err 0.7256 || VAL Loss 0.8288, Err 0.7681
	68: TRAIN Loss 0.7865,  Err 0.7255 || VAL Loss 0.8290, Err 0.7681
	69: TRAIN Loss 0.7864,  Err 0.7253 || VAL Loss 0.8291, Err 0.7680
	70: TRAIN Loss 0.7864,  Err 0.7252 || VAL Loss 0.8292, Err 0.7680
	71: TRAIN Loss 0.7864,  Err 0.7249 || VAL Loss 0.8293, Err 0.7679
	72: TRAIN Loss 0.7864,  Err 0.7247 || VAL Loss 0.8293, Err 0.7679
	73: TRAIN Loss 0.7864,  Err 0.7246 || VAL Loss 0.8295, Err 0.7679
	74: TRAIN Loss 0.7863,  Err 0.7245 || VAL Loss 0.8296, Err 0.7678
	75: TRAIN Loss 0.7863,  Err 0.7243 || VAL Loss 0.8296, Err 0.7677
	76: TRAIN Loss 0.7863,  Err 0.7242 || VAL Loss 0.8297, Err 0.7677
	77: TRAIN Loss 0.7863,  Err 0.7240 || VAL Loss 0.8298, Err 0.7677
	78: TRAIN Loss 0.7863,  Err 0.7238 || VAL Loss 0.8299, Err 0.7676
	79: TRAIN Loss 0.7862,  Err 0.7238 || VAL Loss 0.8299, Err 0.7676
	80: TRAIN Loss 0.7862,  Err 0.7236 || VAL Loss 0.8300, Err 0.7675
	81: TRAIN Loss 0.7862,  Err 0.7235 || VAL Loss 0.8301, Err 0.

	88: TRAIN Loss 0.8050,  Err 0.8005 || VAL Loss 0.8247, Err 0.8203
	89: TRAIN Loss 0.8050,  Err 0.8005 || VAL Loss 0.8247, Err 0.8203
	90: TRAIN Loss 0.8050,  Err 0.8005 || VAL Loss 0.8247, Err 0.8202
	91: TRAIN Loss 0.8050,  Err 0.8005 || VAL Loss 0.8247, Err 0.8202
	92: TRAIN Loss 0.8050,  Err 0.8004 || VAL Loss 0.8247, Err 0.8202
	93: TRAIN Loss 0.8049,  Err 0.8004 || VAL Loss 0.8247, Err 0.8202
	94: TRAIN Loss 0.8050,  Err 0.8004 || VAL Loss 0.8247, Err 0.8202
	95: TRAIN Loss 0.8049,  Err 0.8004 || VAL Loss 0.8247, Err 0.8202
	96: TRAIN Loss 0.8049,  Err 0.8003 || VAL Loss 0.8247, Err 0.8202
	97: TRAIN Loss 0.8049,  Err 0.8004 || VAL Loss 0.8247, Err 0.8202
	98: TRAIN Loss 0.8049,  Err 0.8003 || VAL Loss 0.8247, Err 0.8201
	99: TRAIN Loss 0.8049,  Err 0.8003 || VAL Loss 0.8247, Err 0.8201

==> 20,0.0001,0.8200971972631319 == best (10,2e-05,0.7366513533519797) =============

F 30, penalty 0 :
	0: TRAIN Loss 1.2267,  Err 1.2267 || VAL Loss 1.1753, Err 1.1753
	1: TRAIN Loss 1.1143,  E

	8: TRAIN Loss 0.7756,  Err 0.7659 || VAL Loss 0.8824, Err 0.8725
	9: TRAIN Loss 0.7491,  Err 0.7389 || VAL Loss 0.8705, Err 0.8601
	10: TRAIN Loss 0.7252,  Err 0.7146 || VAL Loss 0.8615, Err 0.8506
	11: TRAIN Loss 0.7038,  Err 0.6925 || VAL Loss 0.8550, Err 0.8434
	12: TRAIN Loss 0.6843,  Err 0.6724 || VAL Loss 0.8505, Err 0.8383
	13: TRAIN Loss 0.6665,  Err 0.6540 || VAL Loss 0.8477, Err 0.8348
	14: TRAIN Loss 0.6502,  Err 0.6370 || VAL Loss 0.8463, Err 0.8328
	15: TRAIN Loss 0.6352,  Err 0.6213 || VAL Loss 0.8459, Err 0.8317
	16: TRAIN Loss 0.6213,  Err 0.6069 || VAL Loss 0.8464, Err 0.8316
	17: TRAIN Loss 0.6086,  Err 0.5935 || VAL Loss 0.8476, Err 0.8322
	18: TRAIN Loss 0.5968,  Err 0.5811 || VAL Loss 0.8494, Err 0.8334
	19: TRAIN Loss 0.5860,  Err 0.5697 || VAL Loss 0.8516, Err 0.8349
	20: TRAIN Loss 0.5761,  Err 0.5591 || VAL Loss 0.8541, Err 0.8369
	21: TRAIN Loss 0.5669,  Err 0.5493 || VAL Loss 0.8568, Err 0.8390
	22: TRAIN Loss 0.5584,  Err 0.5403 || VAL Loss 0.8596, Err 0.84

	29: TRAIN Loss 0.5796,  Err 0.5103 || VAL Loss 0.8800, Err 0.8100
	30: TRAIN Loss 0.5758,  Err 0.5053 || VAL Loss 0.8823, Err 0.8112
	31: TRAIN Loss 0.5723,  Err 0.5006 || VAL Loss 0.8845, Err 0.8123
	32: TRAIN Loss 0.5689,  Err 0.4962 || VAL Loss 0.8867, Err 0.8135
	33: TRAIN Loss 0.5659,  Err 0.4921 || VAL Loss 0.8889, Err 0.8147
	34: TRAIN Loss 0.5629,  Err 0.4882 || VAL Loss 0.8909, Err 0.8157
	35: TRAIN Loss 0.5602,  Err 0.4845 || VAL Loss 0.8929, Err 0.8168
	36: TRAIN Loss 0.5576,  Err 0.4811 || VAL Loss 0.8949, Err 0.8179
	37: TRAIN Loss 0.5552,  Err 0.4778 || VAL Loss 0.8968, Err 0.8190
	38: TRAIN Loss 0.5529,  Err 0.4747 || VAL Loss 0.8985, Err 0.8200
	39: TRAIN Loss 0.5507,  Err 0.4718 || VAL Loss 0.9002, Err 0.8209
	40: TRAIN Loss 0.5487,  Err 0.4690 || VAL Loss 0.9020, Err 0.8219
	41: TRAIN Loss 0.5467,  Err 0.4663 || VAL Loss 0.9035, Err 0.8228
	42: TRAIN Loss 0.5449,  Err 0.4638 || VAL Loss 0.9050, Err 0.8237
	43: TRAIN Loss 0.5431,  Err 0.4614 || VAL Loss 0.9066, Err 0.

	50: TRAIN Loss 0.5989,  Err 0.4812 || VAL Loss 0.9066, Err 0.7887
	51: TRAIN Loss 0.5980,  Err 0.4797 || VAL Loss 0.9073, Err 0.7888
	52: TRAIN Loss 0.5971,  Err 0.4783 || VAL Loss 0.9081, Err 0.7891
	53: TRAIN Loss 0.5963,  Err 0.4769 || VAL Loss 0.9089, Err 0.7893
	54: TRAIN Loss 0.5955,  Err 0.4756 || VAL Loss 0.9097, Err 0.7895
	55: TRAIN Loss 0.5948,  Err 0.4744 || VAL Loss 0.9103, Err 0.7897
	56: TRAIN Loss 0.5941,  Err 0.4732 || VAL Loss 0.9110, Err 0.7899
	57: TRAIN Loss 0.5934,  Err 0.4720 || VAL Loss 0.9117, Err 0.7902
	58: TRAIN Loss 0.5927,  Err 0.4709 || VAL Loss 0.9123, Err 0.7903
	59: TRAIN Loss 0.5921,  Err 0.4699 || VAL Loss 0.9129, Err 0.7906
	60: TRAIN Loss 0.5915,  Err 0.4689 || VAL Loss 0.9136, Err 0.7908
	61: TRAIN Loss 0.5909,  Err 0.4678 || VAL Loss 0.9141, Err 0.7909
	62: TRAIN Loss 0.5903,  Err 0.4669 || VAL Loss 0.9146, Err 0.7911
	63: TRAIN Loss 0.5898,  Err 0.4660 || VAL Loss 0.9152, Err 0.7913
	64: TRAIN Loss 0.5892,  Err 0.4651 || VAL Loss 0.9158, Err 0.

	71: TRAIN Loss 0.6777,  Err 0.5337 || VAL Loss 0.8918, Err 0.7477
	72: TRAIN Loss 0.6775,  Err 0.5332 || VAL Loss 0.8921, Err 0.7478
	73: TRAIN Loss 0.6773,  Err 0.5327 || VAL Loss 0.8925, Err 0.7478
	74: TRAIN Loss 0.6771,  Err 0.5323 || VAL Loss 0.8926, Err 0.7477
	75: TRAIN Loss 0.6770,  Err 0.5319 || VAL Loss 0.8929, Err 0.7478
	76: TRAIN Loss 0.6768,  Err 0.5314 || VAL Loss 0.8932, Err 0.7478
	77: TRAIN Loss 0.6766,  Err 0.5310 || VAL Loss 0.8935, Err 0.7479
	78: TRAIN Loss 0.6764,  Err 0.5306 || VAL Loss 0.8936, Err 0.7478
	79: TRAIN Loss 0.6763,  Err 0.5302 || VAL Loss 0.8939, Err 0.7478
	80: TRAIN Loss 0.6761,  Err 0.5298 || VAL Loss 0.8941, Err 0.7478
	81: TRAIN Loss 0.6760,  Err 0.5295 || VAL Loss 0.8943, Err 0.7478
	82: TRAIN Loss 0.6758,  Err 0.5291 || VAL Loss 0.8945, Err 0.7478
	83: TRAIN Loss 0.6757,  Err 0.5288 || VAL Loss 0.8948, Err 0.7479
	84: TRAIN Loss 0.6755,  Err 0.5284 || VAL Loss 0.8949, Err 0.7479
	85: TRAIN Loss 0.6754,  Err 0.5281 || VAL Loss 0.8952, Err 0.

	92: TRAIN Loss 0.7319,  Err 0.5965 || VAL Loss 0.8738, Err 0.7384
	93: TRAIN Loss 0.7318,  Err 0.5963 || VAL Loss 0.8739, Err 0.7384
	94: TRAIN Loss 0.7318,  Err 0.5961 || VAL Loss 0.8741, Err 0.7385
	95: TRAIN Loss 0.7318,  Err 0.5960 || VAL Loss 0.8742, Err 0.7385
	96: TRAIN Loss 0.7317,  Err 0.5958 || VAL Loss 0.8742, Err 0.7384
	97: TRAIN Loss 0.7317,  Err 0.5956 || VAL Loss 0.8744, Err 0.7384
	98: TRAIN Loss 0.7316,  Err 0.5955 || VAL Loss 0.8745, Err 0.7384
	99: TRAIN Loss 0.7316,  Err 0.5953 || VAL Loss 0.8746, Err 0.7384

==> 30,3e-05,0.7384152587478772 == best (10,2e-05,0.7366513533519797) =============

F 30, penalty 5e-05 :
	0: TRAIN Loss 1.5935,  Err 1.2191 || VAL Loss 1.4493, Err 1.1641
	1: TRAIN Loss 1.3321,  Err 1.1057 || VAL Loss 1.2557, Err 1.0797
	2: TRAIN Loss 1.1697,  Err 1.0279 || VAL Loss 1.1311, Err 1.0187
	3: TRAIN Loss 1.0652,  Err 0.9729 || VAL Loss 1.0487, Err 0.9739
	4: TRAIN Loss 0.9960,  Err 0.9332 || VAL Loss 0.9925, Err 0.9403
	5: TRAIN Loss 0.9487,  Er

	12: TRAIN Loss 0.8377,  Err 0.8336 || VAL Loss 0.8498, Err 0.8458
	13: TRAIN Loss 0.8331,  Err 0.8289 || VAL Loss 0.8460, Err 0.8419
	14: TRAIN Loss 0.8293,  Err 0.8250 || VAL Loss 0.8429, Err 0.8387
	15: TRAIN Loss 0.8260,  Err 0.8217 || VAL Loss 0.8403, Err 0.8360
	16: TRAIN Loss 0.8233,  Err 0.8189 || VAL Loss 0.8381, Err 0.8337
	17: TRAIN Loss 0.8210,  Err 0.8165 || VAL Loss 0.8362, Err 0.8317
	18: TRAIN Loss 0.8190,  Err 0.8144 || VAL Loss 0.8346, Err 0.8301
	19: TRAIN Loss 0.8174,  Err 0.8127 || VAL Loss 0.8333, Err 0.8287
	20: TRAIN Loss 0.8159,  Err 0.8112 || VAL Loss 0.8322, Err 0.8275
	21: TRAIN Loss 0.8147,  Err 0.8100 || VAL Loss 0.8312, Err 0.8265
	22: TRAIN Loss 0.8136,  Err 0.8088 || VAL Loss 0.8303, Err 0.8257
	23: TRAIN Loss 0.8126,  Err 0.8079 || VAL Loss 0.8296, Err 0.8249
	24: TRAIN Loss 0.8118,  Err 0.8071 || VAL Loss 0.8289, Err 0.8243
	25: TRAIN Loss 0.8111,  Err 0.8064 || VAL Loss 0.8284, Err 0.8238
	26: TRAIN Loss 0.8105,  Err 0.8058 || VAL Loss 0.8279, Err 0.

	33: TRAIN Loss 0.3470,  Err 0.3470 || VAL Loss 0.9615, Err 0.9615
	34: TRAIN Loss 0.3427,  Err 0.3427 || VAL Loss 0.9671, Err 0.9671
	35: TRAIN Loss 0.3387,  Err 0.3387 || VAL Loss 0.9725, Err 0.9725
	36: TRAIN Loss 0.3349,  Err 0.3349 || VAL Loss 0.9778, Err 0.9778
	37: TRAIN Loss 0.3312,  Err 0.3312 || VAL Loss 0.9831, Err 0.9831
	38: TRAIN Loss 0.3278,  Err 0.3278 || VAL Loss 0.9883, Err 0.9883
	39: TRAIN Loss 0.3245,  Err 0.3245 || VAL Loss 0.9935, Err 0.9935
	40: TRAIN Loss 0.3213,  Err 0.3213 || VAL Loss 0.9986, Err 0.9986
	41: TRAIN Loss 0.3183,  Err 0.3183 || VAL Loss 1.0036, Err 1.0036
	42: TRAIN Loss 0.3155,  Err 0.3155 || VAL Loss 1.0087, Err 1.0087
	43: TRAIN Loss 0.3127,  Err 0.3127 || VAL Loss 1.0136, Err 1.0136
	44: TRAIN Loss 0.3101,  Err 0.3101 || VAL Loss 1.0184, Err 1.0184
	45: TRAIN Loss 0.3076,  Err 0.3076 || VAL Loss 1.0233, Err 1.0233
	46: TRAIN Loss 0.3051,  Err 0.3051 || VAL Loss 1.0281, Err 1.0281
	47: TRAIN Loss 0.3028,  Err 0.3028 || VAL Loss 1.0329, Err 1.

	54: TRAIN Loss 0.3337,  Err 0.2952 || VAL Loss 1.0447, Err 1.0060
	55: TRAIN Loss 0.3323,  Err 0.2936 || VAL Loss 1.0480, Err 1.0092
	56: TRAIN Loss 0.3308,  Err 0.2919 || VAL Loss 1.0513, Err 1.0122
	57: TRAIN Loss 0.3295,  Err 0.2903 || VAL Loss 1.0547, Err 1.0154
	58: TRAIN Loss 0.3281,  Err 0.2888 || VAL Loss 1.0580, Err 1.0186
	59: TRAIN Loss 0.3268,  Err 0.2874 || VAL Loss 1.0610, Err 1.0215
	60: TRAIN Loss 0.3256,  Err 0.2859 || VAL Loss 1.0643, Err 1.0245
	61: TRAIN Loss 0.3244,  Err 0.2845 || VAL Loss 1.0673, Err 1.0274
	62: TRAIN Loss 0.3232,  Err 0.2832 || VAL Loss 1.0704, Err 1.0303
	63: TRAIN Loss 0.3221,  Err 0.2819 || VAL Loss 1.0733, Err 1.0330
	64: TRAIN Loss 0.3210,  Err 0.2806 || VAL Loss 1.0764, Err 1.0359
	65: TRAIN Loss 0.3199,  Err 0.2794 || VAL Loss 1.0792, Err 1.0386
	66: TRAIN Loss 0.3189,  Err 0.2782 || VAL Loss 1.0822, Err 1.0414
	67: TRAIN Loss 0.3179,  Err 0.2771 || VAL Loss 1.0850, Err 1.0441
	68: TRAIN Loss 0.3169,  Err 0.2759 || VAL Loss 1.0877, Err 1.

	75: TRAIN Loss 0.4311,  Err 0.3050 || VAL Loss 1.0475, Err 0.9212
	76: TRAIN Loss 0.4306,  Err 0.3042 || VAL Loss 1.0486, Err 0.9221
	77: TRAIN Loss 0.4301,  Err 0.3035 || VAL Loss 1.0498, Err 0.9231
	78: TRAIN Loss 0.4296,  Err 0.3028 || VAL Loss 1.0509, Err 0.9241
	79: TRAIN Loss 0.4291,  Err 0.3021 || VAL Loss 1.0519, Err 0.9249
	80: TRAIN Loss 0.4286,  Err 0.3014 || VAL Loss 1.0530, Err 0.9258
	81: TRAIN Loss 0.4281,  Err 0.3008 || VAL Loss 1.0540, Err 0.9266
	82: TRAIN Loss 0.4276,  Err 0.3001 || VAL Loss 1.0551, Err 0.9275
	83: TRAIN Loss 0.4271,  Err 0.2995 || VAL Loss 1.0561, Err 0.9284
	84: TRAIN Loss 0.4267,  Err 0.2989 || VAL Loss 1.0570, Err 0.9291
	85: TRAIN Loss 0.4263,  Err 0.2983 || VAL Loss 1.0580, Err 0.9299
	86: TRAIN Loss 0.4258,  Err 0.2977 || VAL Loss 1.0590, Err 0.9308
	87: TRAIN Loss 0.4254,  Err 0.2971 || VAL Loss 1.0599, Err 0.9315
	88: TRAIN Loss 0.4250,  Err 0.2965 || VAL Loss 1.0607, Err 0.9322
	89: TRAIN Loss 0.4246,  Err 0.2960 || VAL Loss 1.0616, Err 0.

	96: TRAIN Loss 0.5236,  Err 0.3474 || VAL Loss 1.0159, Err 0.8396
	97: TRAIN Loss 0.5234,  Err 0.3470 || VAL Loss 1.0163, Err 0.8399
	98: TRAIN Loss 0.5231,  Err 0.3466 || VAL Loss 1.0169, Err 0.8403
	99: TRAIN Loss 0.5229,  Err 0.3462 || VAL Loss 1.0175, Err 0.8408

==> 50,1e-05,0.8407848567520759 == best (10,2e-05,0.7366513533519797) =============

F 50, penalty 2e-05 :
	0: TRAIN Loss 1.3737,  Err 1.2119 || VAL Loss 1.2946, Err 1.1584
	1: TRAIN Loss 1.2157,  Err 1.0968 || VAL Loss 1.1811, Err 1.0778
	2: TRAIN Loss 1.1056,  Err 1.0133 || VAL Loss 1.0990, Err 1.0167
	3: TRAIN Loss 1.0252,  Err 0.9496 || VAL Loss 1.0376, Err 0.9682
	4: TRAIN Loss 0.9639,  Err 0.8983 || VAL Loss 0.9900, Err 0.9277
	5: TRAIN Loss 0.9153,  Err 0.8543 || VAL Loss 0.9523, Err 0.8925
	6: TRAIN Loss 0.8762,  Err 0.8160 || VAL Loss 0.9232, Err 0.8626
	7: TRAIN Loss 0.8457,  Err 0.7836 || VAL Loss 0.9024, Err 0.8389
	8: TRAIN Loss 0.8224,  Err 0.7568 || VAL Loss 0.8882, Err 0.8208
	9: TRAIN Loss 0.8042,  Err 0.

	16: TRAIN Loss 0.7769,  Err 0.6909 || VAL Loss 0.8524, Err 0.7648
	17: TRAIN Loss 0.7724,  Err 0.6827 || VAL Loss 0.8522, Err 0.7610
	18: TRAIN Loss 0.7683,  Err 0.6751 || VAL Loss 0.8524, Err 0.7577
	19: TRAIN Loss 0.7646,  Err 0.6679 || VAL Loss 0.8530, Err 0.7548
	20: TRAIN Loss 0.7613,  Err 0.6612 || VAL Loss 0.8537, Err 0.7523
	21: TRAIN Loss 0.7583,  Err 0.6550 || VAL Loss 0.8548, Err 0.7502
	22: TRAIN Loss 0.7556,  Err 0.6492 || VAL Loss 0.8559, Err 0.7483
	23: TRAIN Loss 0.7532,  Err 0.6438 || VAL Loss 0.8571, Err 0.7466
	24: TRAIN Loss 0.7510,  Err 0.6388 || VAL Loss 0.8584, Err 0.7452
	25: TRAIN Loss 0.7490,  Err 0.6341 || VAL Loss 0.8599, Err 0.7440
	26: TRAIN Loss 0.7472,  Err 0.6298 || VAL Loss 0.8614, Err 0.7430
	27: TRAIN Loss 0.7456,  Err 0.6258 || VAL Loss 0.8627, Err 0.7420
	28: TRAIN Loss 0.7441,  Err 0.6220 || VAL Loss 0.8642, Err 0.7413
	29: TRAIN Loss 0.7428,  Err 0.6185 || VAL Loss 0.8656, Err 0.7406
	30: TRAIN Loss 0.7415,  Err 0.6151 || VAL Loss 0.8670, Err 0.

	37: TRAIN Loss 0.7904,  Err 0.7301 || VAL Loss 0.8293, Err 0.7693
	38: TRAIN Loss 0.7902,  Err 0.7297 || VAL Loss 0.8293, Err 0.7692
	39: TRAIN Loss 0.7900,  Err 0.7293 || VAL Loss 0.8293, Err 0.7690
	40: TRAIN Loss 0.7898,  Err 0.7291 || VAL Loss 0.8294, Err 0.7689
	41: TRAIN Loss 0.7897,  Err 0.7286 || VAL Loss 0.8295, Err 0.7688
	42: TRAIN Loss 0.7895,  Err 0.7283 || VAL Loss 0.8295, Err 0.7687
	43: TRAIN Loss 0.7894,  Err 0.7280 || VAL Loss 0.8296, Err 0.7686
	44: TRAIN Loss 0.7893,  Err 0.7277 || VAL Loss 0.8297, Err 0.7685
	45: TRAIN Loss 0.7892,  Err 0.7275 || VAL Loss 0.8297, Err 0.7683
	46: TRAIN Loss 0.7891,  Err 0.7272 || VAL Loss 0.8298, Err 0.7683
	47: TRAIN Loss 0.7890,  Err 0.7270 || VAL Loss 0.8299, Err 0.7682
	48: TRAIN Loss 0.7889,  Err 0.7267 || VAL Loss 0.8299, Err 0.7682
	49: TRAIN Loss 0.7888,  Err 0.7265 || VAL Loss 0.8300, Err 0.7681
	50: TRAIN Loss 0.7888,  Err 0.7264 || VAL Loss 0.8300, Err 0.7680
	51: TRAIN Loss 0.7887,  Err 0.7262 || VAL Loss 0.8302, Err 0.

	58: TRAIN Loss 0.8054,  Err 0.8012 || VAL Loss 0.8246, Err 0.8205
	59: TRAIN Loss 0.8054,  Err 0.8012 || VAL Loss 0.8246, Err 0.8205
	60: TRAIN Loss 0.8053,  Err 0.8011 || VAL Loss 0.8246, Err 0.8205
	61: TRAIN Loss 0.8053,  Err 0.8011 || VAL Loss 0.8246, Err 0.8204
	62: TRAIN Loss 0.8053,  Err 0.8011 || VAL Loss 0.8246, Err 0.8205
	63: TRAIN Loss 0.8053,  Err 0.8011 || VAL Loss 0.8246, Err 0.8204
	64: TRAIN Loss 0.8052,  Err 0.8010 || VAL Loss 0.8246, Err 0.8204
	65: TRAIN Loss 0.8052,  Err 0.8010 || VAL Loss 0.8246, Err 0.8204
	66: TRAIN Loss 0.8052,  Err 0.8010 || VAL Loss 0.8246, Err 0.8204
	67: TRAIN Loss 0.8052,  Err 0.8009 || VAL Loss 0.8246, Err 0.8204
	68: TRAIN Loss 0.8052,  Err 0.8009 || VAL Loss 0.8247, Err 0.8204
	69: TRAIN Loss 0.8052,  Err 0.8009 || VAL Loss 0.8246, Err 0.8204
	70: TRAIN Loss 0.8052,  Err 0.8009 || VAL Loss 0.8246, Err 0.8204
	71: TRAIN Loss 0.8052,  Err 0.8008 || VAL Loss 0.8246, Err 0.8203
	72: TRAIN Loss 0.8051,  Err 0.8008 || VAL Loss 0.8246, Err 0.

	79: TRAIN Loss 0.0864,  Err 0.0864 || VAL Loss 1.4581, Err 1.4581
	80: TRAIN Loss 0.0856,  Err 0.0856 || VAL Loss 1.4653, Err 1.4653
	81: TRAIN Loss 0.0850,  Err 0.0850 || VAL Loss 1.4726, Err 1.4726
	82: TRAIN Loss 0.0843,  Err 0.0843 || VAL Loss 1.4798, Err 1.4798
	83: TRAIN Loss 0.0837,  Err 0.0837 || VAL Loss 1.4871, Err 1.4871
	84: TRAIN Loss 0.0830,  Err 0.0830 || VAL Loss 1.4944, Err 1.4944
	85: TRAIN Loss 0.0824,  Err 0.0824 || VAL Loss 1.5014, Err 1.5014
	86: TRAIN Loss 0.0818,  Err 0.0818 || VAL Loss 1.5086, Err 1.5086
	87: TRAIN Loss 0.0813,  Err 0.0813 || VAL Loss 1.5157, Err 1.5157
	88: TRAIN Loss 0.0807,  Err 0.0807 || VAL Loss 1.5227, Err 1.5227
	89: TRAIN Loss 0.0801,  Err 0.0801 || VAL Loss 1.5300, Err 1.5300
	90: TRAIN Loss 0.0796,  Err 0.0796 || VAL Loss 1.5368, Err 1.5368
	91: TRAIN Loss 0.0790,  Err 0.0790 || VAL Loss 1.5440, Err 1.5440
	92: TRAIN Loss 0.0785,  Err 0.0785 || VAL Loss 1.5510, Err 1.5510
	93: TRAIN Loss 0.0780,  Err 0.0780 || VAL Loss 1.5579, Err 1.


==> 100,1e-06,1.3321577114443366 == best (10,2e-05,0.7366513533519797) =============

F 100, penalty 5e-06 :
	0: TRAIN Loss 1.2486,  Err 1.2042 || VAL Loss 1.1953, Err 1.1539
	1: TRAIN Loss 1.1196,  Err 1.0798 || VAL Loss 1.1128, Err 1.0743
	2: TRAIN Loss 1.0186,  Err 0.9804 || VAL Loss 1.0453, Err 1.0072
	3: TRAIN Loss 0.9288,  Err 0.8899 || VAL Loss 0.9833, Err 0.9434
	4: TRAIN Loss 0.8454,  Err 0.8034 || VAL Loss 0.9312, Err 0.8870
	5: TRAIN Loss 0.7762,  Err 0.7293 || VAL Loss 0.8982, Err 0.8485
	6: TRAIN Loss 0.7230,  Err 0.6703 || VAL Loss 0.8807, Err 0.8249
	7: TRAIN Loss 0.6799,  Err 0.6208 || VAL Loss 0.8727, Err 0.8105
	8: TRAIN Loss 0.6426,  Err 0.5770 || VAL Loss 0.8707, Err 0.8019
	9: TRAIN Loss 0.6094,  Err 0.5372 || VAL Loss 0.8728, Err 0.7973
	10: TRAIN Loss 0.5798,  Err 0.5009 || VAL Loss 0.8780, Err 0.7958
	11: TRAIN Loss 0.5534,  Err 0.4680 || VAL Loss 0.8850, Err 0.7964
	12: TRAIN Loss 0.5301,  Err 0.4384 || VAL Loss 0.8938, Err 0.7990
	13: TRAIN Loss 0.5097,  Err 

	20: TRAIN Loss 0.5361,  Err 0.3572 || VAL Loss 0.9748, Err 0.7937
	21: TRAIN Loss 0.5299,  Err 0.3465 || VAL Loss 0.9824, Err 0.7970
	22: TRAIN Loss 0.5243,  Err 0.3369 || VAL Loss 0.9896, Err 0.8003
	23: TRAIN Loss 0.5193,  Err 0.3280 || VAL Loss 0.9965, Err 0.8036
	24: TRAIN Loss 0.5147,  Err 0.3201 || VAL Loss 1.0030, Err 0.8069
	25: TRAIN Loss 0.5106,  Err 0.3128 || VAL Loss 1.0091, Err 0.8099
	26: TRAIN Loss 0.5068,  Err 0.3061 || VAL Loss 1.0148, Err 0.8129
	27: TRAIN Loss 0.5032,  Err 0.3000 || VAL Loss 1.0203, Err 0.8158
	28: TRAIN Loss 0.5000,  Err 0.2943 || VAL Loss 1.0254, Err 0.8186
	29: TRAIN Loss 0.4971,  Err 0.2891 || VAL Loss 1.0302, Err 0.8212
	30: TRAIN Loss 0.4943,  Err 0.2843 || VAL Loss 1.0348, Err 0.8239
	31: TRAIN Loss 0.4917,  Err 0.2798 || VAL Loss 1.0392, Err 0.8264
	32: TRAIN Loss 0.4893,  Err 0.2757 || VAL Loss 1.0433, Err 0.8289
	33: TRAIN Loss 0.4871,  Err 0.2719 || VAL Loss 1.0471, Err 0.8311
	34: TRAIN Loss 0.4850,  Err 0.2682 || VAL Loss 1.0507, Err 0.

	41: TRAIN Loss 0.6366,  Err 0.3962 || VAL Loss 0.9926, Err 0.7518
	42: TRAIN Loss 0.6358,  Err 0.3942 || VAL Loss 0.9941, Err 0.7522
	43: TRAIN Loss 0.6351,  Err 0.3924 || VAL Loss 0.9956, Err 0.7527
	44: TRAIN Loss 0.6344,  Err 0.3906 || VAL Loss 0.9970, Err 0.7530
	45: TRAIN Loss 0.6338,  Err 0.3892 || VAL Loss 0.9984, Err 0.7534
	46: TRAIN Loss 0.6332,  Err 0.3874 || VAL Loss 0.9998, Err 0.7539
	47: TRAIN Loss 0.6327,  Err 0.3861 || VAL Loss 1.0009, Err 0.7542
	48: TRAIN Loss 0.6322,  Err 0.3848 || VAL Loss 1.0021, Err 0.7546
	49: TRAIN Loss 0.6317,  Err 0.3835 || VAL Loss 1.0033, Err 0.7549
	50: TRAIN Loss 0.6312,  Err 0.3823 || VAL Loss 1.0043, Err 0.7552
	51: TRAIN Loss 0.6308,  Err 0.3810 || VAL Loss 1.0054, Err 0.7555
	52: TRAIN Loss 0.6304,  Err 0.3800 || VAL Loss 1.0063, Err 0.7558
	53: TRAIN Loss 0.6300,  Err 0.3789 || VAL Loss 1.0073, Err 0.7562
	54: TRAIN Loss 0.6296,  Err 0.3779 || VAL Loss 1.0082, Err 0.7565
	55: TRAIN Loss 0.6292,  Err 0.3770 || VAL Loss 1.0091, Err 0.

	62: TRAIN Loss 0.7225,  Err 0.5214 || VAL Loss 0.9375, Err 0.7365
	63: TRAIN Loss 0.7224,  Err 0.5208 || VAL Loss 0.9380, Err 0.7366
	64: TRAIN Loss 0.7223,  Err 0.5202 || VAL Loss 0.9385, Err 0.7366
	65: TRAIN Loss 0.7222,  Err 0.5196 || VAL Loss 0.9391, Err 0.7367
	66: TRAIN Loss 0.7221,  Err 0.5190 || VAL Loss 0.9395, Err 0.7366
	67: TRAIN Loss 0.7220,  Err 0.5185 || VAL Loss 0.9401, Err 0.7368
	68: TRAIN Loss 0.7219,  Err 0.5180 || VAL Loss 0.9405, Err 0.7369
	69: TRAIN Loss 0.7219,  Err 0.5175 || VAL Loss 0.9409, Err 0.7369
	70: TRAIN Loss 0.7218,  Err 0.5172 || VAL Loss 0.9413, Err 0.7369
	71: TRAIN Loss 0.7217,  Err 0.5167 || VAL Loss 0.9419, Err 0.7370
	72: TRAIN Loss 0.7216,  Err 0.5161 || VAL Loss 0.9422, Err 0.7370
	73: TRAIN Loss 0.7216,  Err 0.5158 || VAL Loss 0.9426, Err 0.7371
	74: TRAIN Loss 0.7215,  Err 0.5154 || VAL Loss 0.9430, Err 0.7372
	75: TRAIN Loss 0.7214,  Err 0.5150 || VAL Loss 0.9433, Err 0.7372
	76: TRAIN Loss 0.7214,  Err 0.5146 || VAL Loss 0.9436, Err 0.

	83: TRAIN Loss 0.7893,  Err 0.7258 || VAL Loss 0.8310, Err 0.7681
	84: TRAIN Loss 0.7893,  Err 0.7258 || VAL Loss 0.8311, Err 0.7681
	85: TRAIN Loss 0.7893,  Err 0.7257 || VAL Loss 0.8311, Err 0.7682
	86: TRAIN Loss 0.7892,  Err 0.7257 || VAL Loss 0.8312, Err 0.7682
	87: TRAIN Loss 0.7892,  Err 0.7257 || VAL Loss 0.8312, Err 0.7682
	88: TRAIN Loss 0.7892,  Err 0.7256 || VAL Loss 0.8312, Err 0.7682
	89: TRAIN Loss 0.7892,  Err 0.7255 || VAL Loss 0.8312, Err 0.7682
	90: TRAIN Loss 0.7892,  Err 0.7255 || VAL Loss 0.8312, Err 0.7682
	91: TRAIN Loss 0.7891,  Err 0.7255 || VAL Loss 0.8312, Err 0.7682
	92: TRAIN Loss 0.7891,  Err 0.7255 || VAL Loss 0.8312, Err 0.7681
	93: TRAIN Loss 0.7891,  Err 0.7253 || VAL Loss 0.8312, Err 0.7682
	94: TRAIN Loss 0.7890,  Err 0.7254 || VAL Loss 0.8312, Err 0.7681
	95: TRAIN Loss 0.7890,  Err 0.7254 || VAL Loss 0.8313, Err 0.7681
	96: TRAIN Loss 0.7890,  Err 0.7253 || VAL Loss 0.8313, Err 0.7680
	97: TRAIN Loss 0.7890,  Err 0.7252 || VAL Loss 0.8314, Err 0.

	3: TRAIN Loss 0.8049,  Err 0.8049 || VAL Loss 0.8917, Err 0.8917
	4: TRAIN Loss 0.6969,  Err 0.6969 || VAL Loss 0.8461, Err 0.8461
	5: TRAIN Loss 0.6145,  Err 0.6145 || VAL Loss 0.8240, Err 0.8240
	6: TRAIN Loss 0.5444,  Err 0.5444 || VAL Loss 0.8146, Err 0.8146
	7: TRAIN Loss 0.4815,  Err 0.4815 || VAL Loss 0.8132, Err 0.8132
	8: TRAIN Loss 0.4252,  Err 0.4252 || VAL Loss 0.8177, Err 0.8177
	9: TRAIN Loss 0.3758,  Err 0.3758 || VAL Loss 0.8268, Err 0.8268
	10: TRAIN Loss 0.3334,  Err 0.3334 || VAL Loss 0.8388, Err 0.8388
	11: TRAIN Loss 0.2974,  Err 0.2974 || VAL Loss 0.8530, Err 0.8530
	12: TRAIN Loss 0.2670,  Err 0.2670 || VAL Loss 0.8684, Err 0.8684
	13: TRAIN Loss 0.2414,  Err 0.2414 || VAL Loss 0.8841, Err 0.8841
	14: TRAIN Loss 0.2197,  Err 0.2197 || VAL Loss 0.9001, Err 0.9001
	15: TRAIN Loss 0.2011,  Err 0.2011 || VAL Loss 0.9158, Err 0.9158
	16: TRAIN Loss 0.1851,  Err 0.1851 || VAL Loss 0.9314, Err 0.9314
	17: TRAIN Loss 0.1712,  Err 0.1712 || VAL Loss 0.9463, Err 0.9463
	1

	24: TRAIN Loss 0.1642,  Err 0.1155 || VAL Loss 1.0477, Err 0.9987
	25: TRAIN Loss 0.1589,  Err 0.1097 || VAL Loss 1.0582, Err 1.0087
	26: TRAIN Loss 0.1541,  Err 0.1044 || VAL Loss 1.0684, Err 1.0184
	27: TRAIN Loss 0.1497,  Err 0.0996 || VAL Loss 1.0782, Err 1.0278
	28: TRAIN Loss 0.1457,  Err 0.0951 || VAL Loss 1.0879, Err 1.0370
	29: TRAIN Loss 0.1420,  Err 0.0910 || VAL Loss 1.0970, Err 1.0458
	30: TRAIN Loss 0.1385,  Err 0.0872 || VAL Loss 1.1061, Err 1.0545
	31: TRAIN Loss 0.1353,  Err 0.0836 || VAL Loss 1.1148, Err 1.0629
	32: TRAIN Loss 0.1324,  Err 0.0803 || VAL Loss 1.1233, Err 1.0711
	33: TRAIN Loss 0.1296,  Err 0.0773 || VAL Loss 1.1316, Err 1.0791
	34: TRAIN Loss 0.1270,  Err 0.0744 || VAL Loss 1.1397, Err 1.0870
	35: TRAIN Loss 0.1246,  Err 0.0717 || VAL Loss 1.1475, Err 1.0945
	36: TRAIN Loss 0.1223,  Err 0.0692 || VAL Loss 1.1552, Err 1.1019
	37: TRAIN Loss 0.1201,  Err 0.0668 || VAL Loss 1.1627, Err 1.1093
	38: TRAIN Loss 0.1181,  Err 0.0646 || VAL Loss 1.1702, Err 1.

	45: TRAIN Loss 0.2791,  Err 0.0939 || VAL Loss 1.1572, Err 0.9717
	46: TRAIN Loss 0.2780,  Err 0.0925 || VAL Loss 1.1603, Err 0.9745
	47: TRAIN Loss 0.2770,  Err 0.0912 || VAL Loss 1.1631, Err 0.9771
	48: TRAIN Loss 0.2760,  Err 0.0900 || VAL Loss 1.1660, Err 0.9798
	49: TRAIN Loss 0.2750,  Err 0.0889 || VAL Loss 1.1686, Err 0.9822
	50: TRAIN Loss 0.2741,  Err 0.0878 || VAL Loss 1.1712, Err 0.9847
	51: TRAIN Loss 0.2732,  Err 0.0867 || VAL Loss 1.1737, Err 0.9869
	52: TRAIN Loss 0.2724,  Err 0.0857 || VAL Loss 1.1762, Err 0.9893
	53: TRAIN Loss 0.2716,  Err 0.0848 || VAL Loss 1.1785, Err 0.9915
	54: TRAIN Loss 0.2708,  Err 0.0839 || VAL Loss 1.1809, Err 0.9937
	55: TRAIN Loss 0.2700,  Err 0.0830 || VAL Loss 1.1832, Err 0.9959
	56: TRAIN Loss 0.2693,  Err 0.0821 || VAL Loss 1.1854, Err 0.9980
	57: TRAIN Loss 0.2686,  Err 0.0813 || VAL Loss 1.1875, Err 1.0000
	58: TRAIN Loss 0.2680,  Err 0.0806 || VAL Loss 1.1896, Err 1.0020
	59: TRAIN Loss 0.2674,  Err 0.0799 || VAL Loss 1.1916, Err 1.

	66: TRAIN Loss 0.4180,  Err 0.1495 || VAL Loss 1.1393, Err 0.8706
	67: TRAIN Loss 0.4176,  Err 0.1490 || VAL Loss 1.1400, Err 0.8712
	68: TRAIN Loss 0.4173,  Err 0.1485 || VAL Loss 1.1406, Err 0.8717
	69: TRAIN Loss 0.4170,  Err 0.1481 || VAL Loss 1.1414, Err 0.8723
	70: TRAIN Loss 0.4167,  Err 0.1476 || VAL Loss 1.1422, Err 0.8729
	71: TRAIN Loss 0.4164,  Err 0.1471 || VAL Loss 1.1428, Err 0.8734
	72: TRAIN Loss 0.4161,  Err 0.1468 || VAL Loss 1.1434, Err 0.8739
	73: TRAIN Loss 0.4158,  Err 0.1463 || VAL Loss 1.1441, Err 0.8745
	74: TRAIN Loss 0.4156,  Err 0.1459 || VAL Loss 1.1449, Err 0.8751
	75: TRAIN Loss 0.4153,  Err 0.1456 || VAL Loss 1.1454, Err 0.8755
	76: TRAIN Loss 0.4151,  Err 0.1452 || VAL Loss 1.1460, Err 0.8760
	77: TRAIN Loss 0.4148,  Err 0.1448 || VAL Loss 1.1465, Err 0.8764
	78: TRAIN Loss 0.4146,  Err 0.1445 || VAL Loss 1.1470, Err 0.8768
	79: TRAIN Loss 0.4143,  Err 0.1441 || VAL Loss 1.1477, Err 0.8774
	80: TRAIN Loss 0.4141,  Err 0.1439 || VAL Loss 1.1483, Err 0.

	87: TRAIN Loss 0.6132,  Err 0.3174 || VAL Loss 1.0534, Err 0.7577
	88: TRAIN Loss 0.6131,  Err 0.3171 || VAL Loss 1.0535, Err 0.7577
	89: TRAIN Loss 0.6130,  Err 0.3169 || VAL Loss 1.0537, Err 0.7578
	90: TRAIN Loss 0.6129,  Err 0.3168 || VAL Loss 1.0540, Err 0.7579
	91: TRAIN Loss 0.6129,  Err 0.3165 || VAL Loss 1.0543, Err 0.7580
	92: TRAIN Loss 0.6128,  Err 0.3163 || VAL Loss 1.0544, Err 0.7581
	93: TRAIN Loss 0.6127,  Err 0.3161 || VAL Loss 1.0546, Err 0.7581
	94: TRAIN Loss 0.6127,  Err 0.3159 || VAL Loss 1.0547, Err 0.7581
	95: TRAIN Loss 0.6126,  Err 0.3158 || VAL Loss 1.0549, Err 0.7582
	96: TRAIN Loss 0.6125,  Err 0.3156 || VAL Loss 1.0550, Err 0.7583
	97: TRAIN Loss 0.6125,  Err 0.3154 || VAL Loss 1.0551, Err 0.7582
	98: TRAIN Loss 0.6125,  Err 0.3153 || VAL Loss 1.0554, Err 0.7584
	99: TRAIN Loss 0.6124,  Err 0.3151 || VAL Loss 1.0555, Err 0.7584

==> 150,2e-05,0.7584373678137889 == best (10,2e-05,0.7366513533519797) =============

F 150, penalty 3e-05 :
	0: TRAIN Loss 1.39

	7: TRAIN Loss 0.8719,  Err 0.8028 || VAL Loss 0.8888, Err 0.8208
	8: TRAIN Loss 0.8616,  Err 0.7931 || VAL Loss 0.8801, Err 0.8131
	9: TRAIN Loss 0.8526,  Err 0.7851 || VAL Loss 0.8729, Err 0.8068
	10: TRAIN Loss 0.8447,  Err 0.7779 || VAL Loss 0.8667, Err 0.8012
	11: TRAIN Loss 0.8378,  Err 0.7717 || VAL Loss 0.8613, Err 0.7963
	12: TRAIN Loss 0.8319,  Err 0.7663 || VAL Loss 0.8568, Err 0.7923
	13: TRAIN Loss 0.8268,  Err 0.7615 || VAL Loss 0.8529, Err 0.7888
	14: TRAIN Loss 0.8223,  Err 0.7575 || VAL Loss 0.8495, Err 0.7858
	15: TRAIN Loss 0.8184,  Err 0.7539 || VAL Loss 0.8467, Err 0.7836
	16: TRAIN Loss 0.8151,  Err 0.7511 || VAL Loss 0.8442, Err 0.7815
	17: TRAIN Loss 0.8122,  Err 0.7486 || VAL Loss 0.8422, Err 0.7797
	18: TRAIN Loss 0.8097,  Err 0.7463 || VAL Loss 0.8405, Err 0.7782
	19: TRAIN Loss 0.8075,  Err 0.7446 || VAL Loss 0.8390, Err 0.7770
	20: TRAIN Loss 0.8056,  Err 0.7429 || VAL Loss 0.8377, Err 0.7757
	21: TRAIN Loss 0.8040,  Err 0.7413 || VAL Loss 0.8364, Err 0.774

	28: TRAIN Loss 0.8097,  Err 0.8056 || VAL Loss 0.8273, Err 0.8233
	29: TRAIN Loss 0.8093,  Err 0.8052 || VAL Loss 0.8269, Err 0.8230
	30: TRAIN Loss 0.8089,  Err 0.8049 || VAL Loss 0.8266, Err 0.8228
	31: TRAIN Loss 0.8085,  Err 0.8046 || VAL Loss 0.8264, Err 0.8226
	32: TRAIN Loss 0.8082,  Err 0.8043 || VAL Loss 0.8262, Err 0.8225
	33: TRAIN Loss 0.8079,  Err 0.8041 || VAL Loss 0.8260, Err 0.8223
	34: TRAIN Loss 0.8077,  Err 0.8039 || VAL Loss 0.8258, Err 0.8221
	35: TRAIN Loss 0.8074,  Err 0.8037 || VAL Loss 0.8256, Err 0.8220
	36: TRAIN Loss 0.8073,  Err 0.8035 || VAL Loss 0.8255, Err 0.8218
	37: TRAIN Loss 0.8071,  Err 0.8033 || VAL Loss 0.8254, Err 0.8218
	38: TRAIN Loss 0.8069,  Err 0.8032 || VAL Loss 0.8253, Err 0.8217
	39: TRAIN Loss 0.8067,  Err 0.8031 || VAL Loss 0.8252, Err 0.8216
	40: TRAIN Loss 0.8066,  Err 0.8030 || VAL Loss 0.8251, Err 0.8215
	41: TRAIN Loss 0.8065,  Err 0.8028 || VAL Loss 0.8250, Err 0.8215
	42: TRAIN Loss 0.8064,  Err 0.8027 || VAL Loss 0.8250, Err 0.

In [80]:
Y_pred.shape

(127527,)

In [81]:
fits=pd.DataFrame(results,columns=["F","penalty","val_rms"])
fits.head()

F   penalty   val_rms
0  1  0.000000  0.810363
1  1  0.000001  0.785315
2  1  0.000005  0.784891
3  1  0.000010  0.783396
4  1  0.000020  0.786322

In [82]:
summary=pd.pivot_table(fits,index="F",columns="penalty",values="val_rms")
summary

penalty  0.000000  0.000001  0.000005  0.000010  0.000020  0.000030  0.000050  \
F                                                                               
1        0.810363  0.785315  0.784891  0.783396  0.786322  0.789517  0.797876   
5        0.786196  0.774025  0.758775  0.742496  0.747487  0.755300  0.774091   
10       0.824282  0.800488  0.763996  0.745207  0.736651  0.742129  0.768487   
15       0.880004  0.841870  0.783684  0.754371  0.737541  0.740076  0.767632   
20       0.931815  0.880725  0.807047  0.765119  0.739928  0.738499  0.767092   
30       1.041985  0.964705  0.851967  0.795744  0.747960  0.738415  0.766855   
50       1.251038  1.113355  0.939751  0.840785  0.759068  0.738797  0.767313   
100      1.598904  1.332158  1.043215  0.895964  0.764677  0.737937  0.768154   
150      1.699842  1.371497  1.053142  0.884558  0.758437  0.736605  0.768550   

penalty  0.000100  
F                  
1        0.820675  
5        0.819684  
10       0.820113  
15       0.820154  
20       0.820097  
30       0.820306  
50       0.820117  
100      0.820301  
150      0.820279

In [83]:
best=fits.iloc[fits["val_rms"].idxmin()]
best

F          150.000000
penalty      0.000030
val_rms      0.736605
Name: 69, dtype: float64

In [84]:
import seaborn as sns

cm = sns.light_palette("#60FF60", reverse=True, as_cmap=True)

s = summary.style.highlight_min(axis=None)
s

In [85]:
F_best=int(best["F"])
penalty_best=best["penalty"]
F_best,penalty_best

(150, 3e-05)

## Cross Validation

In [94]:
nK=5
kfold=KFold(nK,shuffle=True)
folds=[]
for fold in kfold.split(users):
    folds.append(fold)

In [97]:
def ratings_cross_validate(model,users,movies,ratings,folds):
    accuracies=[]
    count=0
    for train,val in folds:
        print()
        print("============= Fold",count+1,"===========")
        users_train=users[train]
        movies_train=movies[train]
        ratings_train=ratings[train]
        mu=ratings_train.mean()
        dratings_train=ratings_train-mu
        users_val=users[val]
        movies_val=movies[val]
        ratings_val=ratings[val]
        dratings_val=ratings_val-mu
        model.fit([users_train,movies_train],dratings_train,
          epochs=steps,batch_size=5000,
          callbacks=[ReportCallback(1,True)],
          verbose=0,
          validation_data=([users_val,movies_val],dratings_val))
        Y_pred=mu+model.predict([users_val,movies_val]).ravel()
        loss=np.mean((ratings_val-Y_pred)**2)
        
        accuracies.append(loss)
        print("======= fold",count+1,"loss",loss,"============")
        print()
        count+=1
    accuracies=np.array(accuracies)
    return accuracies.mean()

In [99]:
results=[]
best_loss=1e10
best_F=None
steps=100
best_penalty=2e-5
if True:
  for F in [5,10,15]: 
        print()
        print(f"F {F} :")
        model=collaborative_filter(N_users,N_movies,F,best_penalty)
        model.compile(optimizer=keras.optimizers.Adam(),
              metrics=[mse_err],
              loss='mean_squared_error')
        loss=ratings_cross_validate(model,users,movies,ratings,
                               folds)
        results.append((F,loss))
        if loss<best_loss:
            best_loss=loss
            best_F=F
        print()
        print(f"==> {F},{best_penalty},{loss} == best ({best_F},{best_penalty},{best_loss}) =============")


F 5 :

============= Fold 1 ===========


/home/manel/anaconda3/lib/python3.6/site-packages/tensorflow/python/ops/gradients_impl.py:108: UserWarning: Converting sparse IndexedSlices to a dense Tensor of unknown shape. This may consume a large amount of memory.
  "Converting sparse IndexedSlices to a dense Tensor of unknown shape. "


	0: TRAIN Loss 1.5461,  Err 1.3691 || VAL Loss 1.4544, Err 1.2913
	1: TRAIN Loss 1.3842,  Err 1.2325 || VAL Loss 1.3235, Err 1.1829
	2: TRAIN Loss 1.2642,  Err 1.1331 || VAL Loss 1.2250, Err 1.1031
	3: TRAIN Loss 1.1733,  Err 1.0594 || VAL Loss 1.1498, Err 1.0437
	4: TRAIN Loss 1.1034,  Err 1.0041 || VAL Loss 1.0917, Err 0.9990
	5: TRAIN Loss 1.0489,  Err 0.9621 || VAL Loss 1.0461, Err 0.9650
	6: TRAIN Loss 1.0060,  Err 0.9298 || VAL Loss 1.0101, Err 0.9387
	7: TRAIN Loss 0.9718,  Err 0.9047 || VAL Loss 0.9813, Err 0.9184
	8: TRAIN Loss 0.9442,  Err 0.8850 || VAL Loss 0.9580, Err 0.9023
	9: TRAIN Loss 0.9218,  Err 0.8693 || VAL Loss 0.9390, Err 0.8896
	10: TRAIN Loss 0.9034,  Err 0.8567 || VAL Loss 0.9234, Err 0.8794
	11: TRAIN Loss 0.8881,  Err 0.8464 || VAL Loss 0.9105, Err 0.8711
	12: TRAIN Loss 0.8754,  Err 0.8381 || VAL Loss 0.8997, Err 0.8644
	13: TRAIN Loss 0.8647,  Err 0.8311 || VAL Loss 0.8907, Err 0.8589
	14: TRAIN Loss 0.8557,  Err 0.8253 || VAL Loss 0.8832, Err 0.8543
	15: 

	22: TRAIN Loss 0.7335,  Err 0.6790 || VAL Loss 0.7920, Err 0.7375
	23: TRAIN Loss 0.7334,  Err 0.6788 || VAL Loss 0.7929, Err 0.7382
	24: TRAIN Loss 0.7332,  Err 0.6785 || VAL Loss 0.7937, Err 0.7390
	25: TRAIN Loss 0.7331,  Err 0.6783 || VAL Loss 0.7945, Err 0.7396
	26: TRAIN Loss 0.7330,  Err 0.6781 || VAL Loss 0.7952, Err 0.7403
	27: TRAIN Loss 0.7329,  Err 0.6779 || VAL Loss 0.7959, Err 0.7408
	28: TRAIN Loss 0.7328,  Err 0.6777 || VAL Loss 0.7965, Err 0.7413
	29: TRAIN Loss 0.7327,  Err 0.6775 || VAL Loss 0.7971, Err 0.7418
	30: TRAIN Loss 0.7326,  Err 0.6773 || VAL Loss 0.7976, Err 0.7423
	31: TRAIN Loss 0.7325,  Err 0.6771 || VAL Loss 0.7982, Err 0.7427
	32: TRAIN Loss 0.7325,  Err 0.6770 || VAL Loss 0.7987, Err 0.7432
	33: TRAIN Loss 0.7324,  Err 0.6768 || VAL Loss 0.7991, Err 0.7435
	34: TRAIN Loss 0.7324,  Err 0.6767 || VAL Loss 0.7996, Err 0.7439
	35: TRAIN Loss 0.7323,  Err 0.6766 || VAL Loss 0.8000, Err 0.7443
	36: TRAIN Loss 0.7322,  Err 0.6764 || VAL Loss 0.8004, Err 0.

	44: TRAIN Loss 0.7312,  Err 0.6738 || VAL Loss 0.8006, Err 0.7432
	45: TRAIN Loss 0.7311,  Err 0.6738 || VAL Loss 0.8008, Err 0.7435
	46: TRAIN Loss 0.7311,  Err 0.6737 || VAL Loss 0.8011, Err 0.7437
	47: TRAIN Loss 0.7311,  Err 0.6737 || VAL Loss 0.8014, Err 0.7440
	48: TRAIN Loss 0.7311,  Err 0.6737 || VAL Loss 0.8015, Err 0.7441
	49: TRAIN Loss 0.7311,  Err 0.6736 || VAL Loss 0.8018, Err 0.7443
	50: TRAIN Loss 0.7310,  Err 0.6736 || VAL Loss 0.8020, Err 0.7445
	51: TRAIN Loss 0.7310,  Err 0.6735 || VAL Loss 0.8021, Err 0.7447
	52: TRAIN Loss 0.7310,  Err 0.6735 || VAL Loss 0.8023, Err 0.7448
	53: TRAIN Loss 0.7310,  Err 0.6735 || VAL Loss 0.8025, Err 0.7450
	54: TRAIN Loss 0.7310,  Err 0.6735 || VAL Loss 0.8026, Err 0.7451
	55: TRAIN Loss 0.7309,  Err 0.6734 || VAL Loss 0.8027, Err 0.7452
	56: TRAIN Loss 0.7309,  Err 0.6734 || VAL Loss 0.8029, Err 0.7454
	57: TRAIN Loss 0.7309,  Err 0.6733 || VAL Loss 0.8031, Err 0.7455
	58: TRAIN Loss 0.7309,  Err 0.6733 || VAL Loss 0.8032, Err 0.

	66: TRAIN Loss 0.7306,  Err 0.6726 || VAL Loss 0.8015, Err 0.7435
	67: TRAIN Loss 0.7306,  Err 0.6726 || VAL Loss 0.8016, Err 0.7436
	68: TRAIN Loss 0.7306,  Err 0.6726 || VAL Loss 0.8017, Err 0.7437
	69: TRAIN Loss 0.7306,  Err 0.6726 || VAL Loss 0.8018, Err 0.7438
	70: TRAIN Loss 0.7306,  Err 0.6725 || VAL Loss 0.8019, Err 0.7439
	71: TRAIN Loss 0.7306,  Err 0.6725 || VAL Loss 0.8020, Err 0.7439
	72: TRAIN Loss 0.7306,  Err 0.6725 || VAL Loss 0.8021, Err 0.7440
	73: TRAIN Loss 0.7306,  Err 0.6726 || VAL Loss 0.8021, Err 0.7441
	74: TRAIN Loss 0.7306,  Err 0.6725 || VAL Loss 0.8021, Err 0.7441
	75: TRAIN Loss 0.7306,  Err 0.6725 || VAL Loss 0.8022, Err 0.7442
	76: TRAIN Loss 0.7305,  Err 0.6725 || VAL Loss 0.8023, Err 0.7442
	77: TRAIN Loss 0.7306,  Err 0.6725 || VAL Loss 0.8023, Err 0.7442
	78: TRAIN Loss 0.7305,  Err 0.6725 || VAL Loss 0.8024, Err 0.7443
	79: TRAIN Loss 0.7305,  Err 0.6725 || VAL Loss 0.8025, Err 0.7444
	80: TRAIN Loss 0.7305,  Err 0.6725 || VAL Loss 0.8025, Err 0.

	88: TRAIN Loss 0.7314,  Err 0.6728 || VAL Loss 0.7978, Err 0.7392
	89: TRAIN Loss 0.7315,  Err 0.6728 || VAL Loss 0.7979, Err 0.7393
	90: TRAIN Loss 0.7314,  Err 0.6728 || VAL Loss 0.7979, Err 0.7393
	91: TRAIN Loss 0.7314,  Err 0.6728 || VAL Loss 0.7980, Err 0.7394
	92: TRAIN Loss 0.7314,  Err 0.6728 || VAL Loss 0.7981, Err 0.7395
	93: TRAIN Loss 0.7314,  Err 0.6728 || VAL Loss 0.7981, Err 0.7394
	94: TRAIN Loss 0.7314,  Err 0.6728 || VAL Loss 0.7981, Err 0.7395
	95: TRAIN Loss 0.7314,  Err 0.6728 || VAL Loss 0.7981, Err 0.7394
	96: TRAIN Loss 0.7314,  Err 0.6727 || VAL Loss 0.7982, Err 0.7395
	97: TRAIN Loss 0.7314,  Err 0.6727 || VAL Loss 0.7982, Err 0.7395
	98: TRAIN Loss 0.7314,  Err 0.6728 || VAL Loss 0.7982, Err 0.7395
	99: TRAIN Loss 0.7314,  Err 0.6727 || VAL Loss 0.7982, Err 0.7395
======= fold 5 loss 0.7395323894772434 ============


==> 5,0.0001,0.7477400302575287 == best (5,2e-05,0.7477400302575287) =============

F 10 :

============= Fold 1 ===========
	0: TRAIN Loss 1.

	7: TRAIN Loss 0.7195,  Err 0.6416 || VAL Loss 0.7586, Err 0.6807
	8: TRAIN Loss 0.7184,  Err 0.6405 || VAL Loss 0.7626, Err 0.6846
	9: TRAIN Loss 0.7174,  Err 0.6394 || VAL Loss 0.7663, Err 0.6881
	10: TRAIN Loss 0.7166,  Err 0.6384 || VAL Loss 0.7697, Err 0.6914
	11: TRAIN Loss 0.7159,  Err 0.6374 || VAL Loss 0.7729, Err 0.6944
	12: TRAIN Loss 0.7153,  Err 0.6366 || VAL Loss 0.7758, Err 0.6971
	13: TRAIN Loss 0.7147,  Err 0.6359 || VAL Loss 0.7785, Err 0.6996
	14: TRAIN Loss 0.7142,  Err 0.6352 || VAL Loss 0.7810, Err 0.7019
	15: TRAIN Loss 0.7138,  Err 0.6345 || VAL Loss 0.7834, Err 0.7040
	16: TRAIN Loss 0.7134,  Err 0.6339 || VAL Loss 0.7855, Err 0.7060
	17: TRAIN Loss 0.7130,  Err 0.6334 || VAL Loss 0.7875, Err 0.7077
	18: TRAIN Loss 0.7127,  Err 0.6328 || VAL Loss 0.7894, Err 0.7095
	19: TRAIN Loss 0.7124,  Err 0.6323 || VAL Loss 0.7911, Err 0.7110
	20: TRAIN Loss 0.7121,  Err 0.6319 || VAL Loss 0.7927, Err 0.7124
	21: TRAIN Loss 0.7118,  Err 0.6314 || VAL Loss 0.7943, Err 0.713

	29: TRAIN Loss 0.7088,  Err 0.6244 || VAL Loss 0.7985, Err 0.7141
	30: TRAIN Loss 0.7087,  Err 0.6243 || VAL Loss 0.7994, Err 0.7149
	31: TRAIN Loss 0.7086,  Err 0.6241 || VAL Loss 0.8002, Err 0.7157
	32: TRAIN Loss 0.7086,  Err 0.6239 || VAL Loss 0.8010, Err 0.7163
	33: TRAIN Loss 0.7085,  Err 0.6238 || VAL Loss 0.8018, Err 0.7171
	34: TRAIN Loss 0.7084,  Err 0.6237 || VAL Loss 0.8025, Err 0.7177
	35: TRAIN Loss 0.7084,  Err 0.6235 || VAL Loss 0.8032, Err 0.7184
	36: TRAIN Loss 0.7083,  Err 0.6234 || VAL Loss 0.8039, Err 0.7190
	37: TRAIN Loss 0.7082,  Err 0.6233 || VAL Loss 0.8046, Err 0.7196
	38: TRAIN Loss 0.7082,  Err 0.6232 || VAL Loss 0.8052, Err 0.7202
	39: TRAIN Loss 0.7081,  Err 0.6230 || VAL Loss 0.8058, Err 0.7207
	40: TRAIN Loss 0.7081,  Err 0.6229 || VAL Loss 0.8064, Err 0.7212
	41: TRAIN Loss 0.7081,  Err 0.6228 || VAL Loss 0.8069, Err 0.7216
	42: TRAIN Loss 0.7080,  Err 0.6227 || VAL Loss 0.8074, Err 0.7221
	43: TRAIN Loss 0.7080,  Err 0.6226 || VAL Loss 0.8079, Err 0.

	51: TRAIN Loss 0.7073,  Err 0.6209 || VAL Loss 0.8100, Err 0.7237
	52: TRAIN Loss 0.7073,  Err 0.6209 || VAL Loss 0.8104, Err 0.7240
	53: TRAIN Loss 0.7072,  Err 0.6208 || VAL Loss 0.8108, Err 0.7244
	54: TRAIN Loss 0.7072,  Err 0.6207 || VAL Loss 0.8110, Err 0.7245
	55: TRAIN Loss 0.7072,  Err 0.6207 || VAL Loss 0.8113, Err 0.7248
	56: TRAIN Loss 0.7072,  Err 0.6206 || VAL Loss 0.8116, Err 0.7251
	57: TRAIN Loss 0.7072,  Err 0.6206 || VAL Loss 0.8119, Err 0.7254
	58: TRAIN Loss 0.7072,  Err 0.6206 || VAL Loss 0.8122, Err 0.7256
	59: TRAIN Loss 0.7071,  Err 0.6206 || VAL Loss 0.8124, Err 0.7258
	60: TRAIN Loss 0.7071,  Err 0.6205 || VAL Loss 0.8126, Err 0.7260
	61: TRAIN Loss 0.7071,  Err 0.6204 || VAL Loss 0.8129, Err 0.7262
	62: TRAIN Loss 0.7071,  Err 0.6204 || VAL Loss 0.8132, Err 0.7265
	63: TRAIN Loss 0.7071,  Err 0.6204 || VAL Loss 0.8133, Err 0.7267
	64: TRAIN Loss 0.7071,  Err 0.6203 || VAL Loss 0.8136, Err 0.7269
	65: TRAIN Loss 0.7071,  Err 0.6203 || VAL Loss 0.8138, Err 0.

	73: TRAIN Loss 0.7077,  Err 0.6202 || VAL Loss 0.8119, Err 0.7244
	74: TRAIN Loss 0.7077,  Err 0.6202 || VAL Loss 0.8120, Err 0.7245
	75: TRAIN Loss 0.7077,  Err 0.6201 || VAL Loss 0.8122, Err 0.7246
	76: TRAIN Loss 0.7077,  Err 0.6201 || VAL Loss 0.8123, Err 0.7248
	77: TRAIN Loss 0.7077,  Err 0.6201 || VAL Loss 0.8123, Err 0.7248
	78: TRAIN Loss 0.7077,  Err 0.6201 || VAL Loss 0.8125, Err 0.7249
	79: TRAIN Loss 0.7077,  Err 0.6201 || VAL Loss 0.8126, Err 0.7251
	80: TRAIN Loss 0.7076,  Err 0.6200 || VAL Loss 0.8128, Err 0.7252
	81: TRAIN Loss 0.7076,  Err 0.6200 || VAL Loss 0.8129, Err 0.7253
	82: TRAIN Loss 0.7076,  Err 0.6200 || VAL Loss 0.8131, Err 0.7254
	83: TRAIN Loss 0.7076,  Err 0.6200 || VAL Loss 0.8132, Err 0.7255
	84: TRAIN Loss 0.7076,  Err 0.6200 || VAL Loss 0.8133, Err 0.7257
	85: TRAIN Loss 0.7076,  Err 0.6199 || VAL Loss 0.8135, Err 0.7258
	86: TRAIN Loss 0.7076,  Err 0.6199 || VAL Loss 0.8135, Err 0.7259
	87: TRAIN Loss 0.7076,  Err 0.6199 || VAL Loss 0.8136, Err 0.

	93: TRAIN Loss 0.6995,  Err 0.5987 || VAL Loss 0.8482, Err 0.7474
	94: TRAIN Loss 0.6993,  Err 0.5984 || VAL Loss 0.8482, Err 0.7472
	95: TRAIN Loss 0.6992,  Err 0.5981 || VAL Loss 0.8482, Err 0.7471
	96: TRAIN Loss 0.6990,  Err 0.5978 || VAL Loss 0.8483, Err 0.7470
	97: TRAIN Loss 0.6989,  Err 0.5975 || VAL Loss 0.8483, Err 0.7469
	98: TRAIN Loss 0.6988,  Err 0.5972 || VAL Loss 0.8484, Err 0.7468
	99: TRAIN Loss 0.6987,  Err 0.5970 || VAL Loss 0.8484, Err 0.7466
======= fold 1 loss 0.7466241203771755 ============


============= Fold 2 ===========
	0: TRAIN Loss 0.7339,  Err 0.6331 || VAL Loss 0.7046, Err 0.6049
	1: TRAIN Loss 0.7270,  Err 0.6280 || VAL Loss 0.7140, Err 0.6157
	2: TRAIN Loss 0.7220,  Err 0.6241 || VAL Loss 0.7227, Err 0.6253
	3: TRAIN Loss 0.7182,  Err 0.6211 || VAL Loss 0.7307, Err 0.6338
	4: TRAIN Loss 0.7153,  Err 0.6186 || VAL Loss 0.7379, Err 0.6414
	5: TRAIN Loss 0.7129,  Err 0.6165 || VAL Loss 0.7446, Err 0.6483
	6: TRAIN Loss 0.7110,  Err 0.6146 || VAL Loss 0

	13: TRAIN Loss 0.6992,  Err 0.5971 || VAL Loss 0.7750, Err 0.6728
	14: TRAIN Loss 0.6987,  Err 0.5963 || VAL Loss 0.7784, Err 0.6760
	15: TRAIN Loss 0.6982,  Err 0.5957 || VAL Loss 0.7814, Err 0.6789
	16: TRAIN Loss 0.6978,  Err 0.5951 || VAL Loss 0.7843, Err 0.6816
	17: TRAIN Loss 0.6974,  Err 0.5945 || VAL Loss 0.7871, Err 0.6841
	18: TRAIN Loss 0.6970,  Err 0.5939 || VAL Loss 0.7895, Err 0.6864
	19: TRAIN Loss 0.6967,  Err 0.5935 || VAL Loss 0.7918, Err 0.6885
	20: TRAIN Loss 0.6964,  Err 0.5930 || VAL Loss 0.7941, Err 0.6906
	21: TRAIN Loss 0.6962,  Err 0.5926 || VAL Loss 0.7961, Err 0.6924
	22: TRAIN Loss 0.6960,  Err 0.5921 || VAL Loss 0.7981, Err 0.6942
	23: TRAIN Loss 0.6957,  Err 0.5918 || VAL Loss 0.7999, Err 0.6959
	24: TRAIN Loss 0.6955,  Err 0.5914 || VAL Loss 0.8017, Err 0.6975
	25: TRAIN Loss 0.6953,  Err 0.5911 || VAL Loss 0.8033, Err 0.6990
	26: TRAIN Loss 0.6952,  Err 0.5908 || VAL Loss 0.8048, Err 0.7004
	27: TRAIN Loss 0.6950,  Err 0.5905 || VAL Loss 0.8063, Err 0.

	35: TRAIN Loss 0.6933,  Err 0.5867 || VAL Loss 0.8133, Err 0.7066
	36: TRAIN Loss 0.6933,  Err 0.5865 || VAL Loss 0.8142, Err 0.7075
	37: TRAIN Loss 0.6932,  Err 0.5863 || VAL Loss 0.8151, Err 0.7083
	38: TRAIN Loss 0.6931,  Err 0.5862 || VAL Loss 0.8159, Err 0.7089
	39: TRAIN Loss 0.6930,  Err 0.5860 || VAL Loss 0.8166, Err 0.7096
	40: TRAIN Loss 0.6930,  Err 0.5859 || VAL Loss 0.8174, Err 0.7103
	41: TRAIN Loss 0.6929,  Err 0.5857 || VAL Loss 0.8181, Err 0.7109
	42: TRAIN Loss 0.6928,  Err 0.5856 || VAL Loss 0.8188, Err 0.7116
	43: TRAIN Loss 0.6928,  Err 0.5855 || VAL Loss 0.8195, Err 0.7121
	44: TRAIN Loss 0.6927,  Err 0.5853 || VAL Loss 0.8201, Err 0.7127
	45: TRAIN Loss 0.6927,  Err 0.5852 || VAL Loss 0.8207, Err 0.7132
	46: TRAIN Loss 0.6926,  Err 0.5851 || VAL Loss 0.8213, Err 0.7138
	47: TRAIN Loss 0.6926,  Err 0.5849 || VAL Loss 0.8219, Err 0.7143
	48: TRAIN Loss 0.6926,  Err 0.5849 || VAL Loss 0.8225, Err 0.7148
	49: TRAIN Loss 0.6925,  Err 0.5848 || VAL Loss 0.8230, Err 0.

	57: TRAIN Loss 0.6929,  Err 0.5841 || VAL Loss 0.8226, Err 0.7138
	58: TRAIN Loss 0.6929,  Err 0.5840 || VAL Loss 0.8229, Err 0.7140
	59: TRAIN Loss 0.6928,  Err 0.5839 || VAL Loss 0.8233, Err 0.7144
	60: TRAIN Loss 0.6928,  Err 0.5839 || VAL Loss 0.8236, Err 0.7147
	61: TRAIN Loss 0.6928,  Err 0.5838 || VAL Loss 0.8239, Err 0.7150
	62: TRAIN Loss 0.6928,  Err 0.5838 || VAL Loss 0.8243, Err 0.7153
	63: TRAIN Loss 0.6928,  Err 0.5837 || VAL Loss 0.8246, Err 0.7155
	64: TRAIN Loss 0.6927,  Err 0.5836 || VAL Loss 0.8250, Err 0.7160
	65: TRAIN Loss 0.6927,  Err 0.5836 || VAL Loss 0.8253, Err 0.7161
	66: TRAIN Loss 0.6927,  Err 0.5835 || VAL Loss 0.8256, Err 0.7164
	67: TRAIN Loss 0.6927,  Err 0.5835 || VAL Loss 0.8258, Err 0.7167
	68: TRAIN Loss 0.6927,  Err 0.5834 || VAL Loss 0.8262, Err 0.7169
	69: TRAIN Loss 0.6926,  Err 0.5834 || VAL Loss 0.8263, Err 0.7171
	70: TRAIN Loss 0.6926,  Err 0.5834 || VAL Loss 0.8266, Err 0.7173
	71: TRAIN Loss 0.6926,  Err 0.5834 || VAL Loss 0.8268, Err 0.

Best model has $F=15$ and $\gamma=2e-5$.

## Test of Best Model

In [ ]:
              
mu=ratings.mean()
dratings=ratings-mu
dratings_test=ratings_test-mu

In [111]:
model=collaborative_filter(N_users,N_movies,best_F,best_penalty)
model.compile(optimizer=keras.optimizers.Adam(),
              metrics=[mse_err],
              loss='mean_squared_error')


In [113]:
model.fit([users,movies],dratings,
          epochs=200,batch_size=5000,
          callbacks=[ReportCallback(1,True)],
          verbose=0,
          validation_data=([users_test,movies_test],dratings_test))

	0: TRAIN Loss 0.7223,  Err 0.6430 || VAL Loss 0.8160, Err 0.7364
	1: TRAIN Loss 0.7215,  Err 0.6414 || VAL Loss 0.8163, Err 0.7358
	2: TRAIN Loss 0.7209,  Err 0.6399 || VAL Loss 0.8165, Err 0.7352
	3: TRAIN Loss 0.7202,  Err 0.6385 || VAL Loss 0.8167, Err 0.7347
	4: TRAIN Loss 0.7197,  Err 0.6372 || VAL Loss 0.8170, Err 0.7342
	5: TRAIN Loss 0.7191,  Err 0.6359 || VAL Loss 0.8172, Err 0.7337
	6: TRAIN Loss 0.7186,  Err 0.6347 || VAL Loss 0.8175, Err 0.7333
	7: TRAIN Loss 0.7181,  Err 0.6336 || VAL Loss 0.8177, Err 0.7329
	8: TRAIN Loss 0.7177,  Err 0.6326 || VAL Loss 0.8179, Err 0.7326
	9: TRAIN Loss 0.7172,  Err 0.6316 || VAL Loss 0.8181, Err 0.7322
	10: TRAIN Loss 0.7169,  Err 0.6307 || VAL Loss 0.8183, Err 0.7319
	11: TRAIN Loss 0.7165,  Err 0.6298 || VAL Loss 0.8184, Err 0.7316
	12: TRAIN Loss 0.7161,  Err 0.6289 || VAL Loss 0.8187, Err 0.7313
	13: TRAIN Loss 0.7158,  Err 0.6281 || VAL Loss 0.8189, Err 0.7310
	14: TRAIN Loss 0.7155,  Err 0.6273 || VAL Loss 0.8190, Err 0.7307
	15: 

	123: TRAIN Loss 0.7068,  Err 0.6073 || VAL Loss 0.8193, Err 0.7199
	124: TRAIN Loss 0.7068,  Err 0.6072 || VAL Loss 0.8193, Err 0.7198
	125: TRAIN Loss 0.7068,  Err 0.6072 || VAL Loss 0.8193, Err 0.7198
	126: TRAIN Loss 0.7068,  Err 0.6072 || VAL Loss 0.8193, Err 0.7198
	127: TRAIN Loss 0.7068,  Err 0.6071 || VAL Loss 0.8193, Err 0.7197
	128: TRAIN Loss 0.7067,  Err 0.6071 || VAL Loss 0.8193, Err 0.7197
	129: TRAIN Loss 0.7067,  Err 0.6070 || VAL Loss 0.8193, Err 0.7197
	130: TRAIN Loss 0.7067,  Err 0.6070 || VAL Loss 0.8192, Err 0.7196
	131: TRAIN Loss 0.7067,  Err 0.6070 || VAL Loss 0.8193, Err 0.7196
	132: TRAIN Loss 0.7067,  Err 0.6070 || VAL Loss 0.8192, Err 0.7196
	133: TRAIN Loss 0.7066,  Err 0.6069 || VAL Loss 0.8192, Err 0.7196
	134: TRAIN Loss 0.7067,  Err 0.6069 || VAL Loss 0.8193, Err 0.7196
	135: TRAIN Loss 0.7066,  Err 0.6069 || VAL Loss 0.8193, Err 0.7196
	136: TRAIN Loss 0.7066,  Err 0.6069 || VAL Loss 0.8193, Err 0.7195
	137: TRAIN Loss 0.7066,  Err 0.6068 || VAL Loss

In [114]:
Y_pred=mu+model.predict([users_test,movies_test]).ravel()
loss=np.mean((ratings_test-Y_pred)**2)
print(best_F,best_penalty,loss)

15 2e-05 0.7187691488387822


Result is comparable to what we obtained with `numpy` (0.726 versus 0.718)